// react-native-minerva/src/components/keyBoardView/type.js
import PropTypes from 'prop-types';

const keyBoardPropTypes = {
    /**
     * Función Callback que se llama cuando se presiona un elemento del teclado.
     */
    addValue: PropTypes.func,
    /**
     * Habilita icono biometria en la celda vacía del teclado.
     */
    biometricAvailable: PropTypes.bool,
    /**
     * Define si el teclado esta habilitado o deshabilitado.
     */
    disabled: PropTypes.bool,
    /**
     * Nombre del icono de biometria.
     */
    iconBiometricsName: PropTypes.string,
    /**
     * Función Callback que se llama cuando se presiona un elemento del teclado.
     */
    onChange: PropTypes.func,
    /**
     * Define el color del teclado.
     */
    white: PropTypes.bool,
};
const keyBoardDefaultProps = {
    addValue: () => null,
    biometricAvailable: false,
    disabled: false,
    iconBiometricsName: 'icon-bch-biometrics',
    onChange: () => null,
    white: false,
};

const KeyBoardIconPropTypes = {
    type: PropTypes.string,
};

const KeyBoardButtonPropTypes = {
    item: PropTypes.shape({
        title: PropTypes.string,
        type: KeyBoardIconPropTypes.type,
    }),
};

const KeyBoardIconDefaultProps = {
    type: '',
};

export {
    KeyBoardButtonPropTypes,
    keyBoardDefaultProps,
    KeyBoardIconDefaultProps,
    KeyBoardIconPropTypes,
    keyBoardPropTypes,
};
