// react-native-minerva/src/components/listItemResumen/types.js
import PropTypes from 'prop-types';

const propTypes = {
    item: PropTypes.shape({
        /**
         * Estilos para el contenedor de la columna izquierda.
         */
        customStyleListItemLeft: PropTypes.oneOfType([
            PropTypes.object,
            PropTypes.array,
        ]),
        /**
         * Estilos para el contenedor de la columna derecha.
         */
        customStyleListItemRight: PropTypes.oneOfType([
            PropTypes.object,
            PropTypes.array,
        ]),
        /**
         * Estilos para el contenedor del title.
         */
        customTitleListItemStyle: PropTypes.oneOfType([
            PropTypes.object,
            PropTypes.array,
        ]),
        /**
         * Boleano para habilitar y deshabilitar la fila.
         */
        disabled: PropTypes.bool,
        /**
         * Texto lado izquierdo del resumen (una columna).
         */
        label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        /**
         * Propiedades del <Text> de label.
         */
        labelProps: PropTypes.objectOf(PropTypes.any),
        /**
         * Componente que reemplaza a label y renderizar un componente en el lado izquierdo del resumen (una columna).
         */
        renderLeft: PropTypes.oneOfType([
            PropTypes.node,
            PropTypes.func,
            PropTypes.bool,
        ]),
        /**
         * Componente que reemplaza a value y renderizar un componente en el lado derecho del resumen (una columna).
         */
        renderRight: PropTypes.oneOfType([
            PropTypes.node,
            PropTypes.func,
            PropTypes.bool,
        ]),
        /**
         * Muestra linea divisora en la parte inferior de la fila.
         */
        separator: PropTypes.bool,
        /**
         * Texto que renderiza arriba de la fila (label y value).
         */
        title: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        /**
         * Texto lado derecho del resumen (una columna).
         */
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        /**
         * Propiedades del <Text> de value.
         */
        valueProps: PropTypes.objectOf(PropTypes.any),
        /**
         * Muestra linea divisora en la parte inferior del title.
         */
        titleSeparator: PropTypes.bool,
    }),
};

const defaultProps = {
    item: {},
};

export { propTypes, defaultProps };
