// react-native-minerva/src/components/logo/logo.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Images from 'bch-assets';
import { Image } from 'react-native';
import Style from './style';
import { HelperOptions } from '../../index';

/**
 * Componente Logos Apps y Marca.
 *
 * @component
 */
const Logo = (props) => {
    const currentBrandName = HelperOptions.brandName.toLowerCase();
    const { logoBanco, logoApps, logoCombined, isotipos } = Images;
    const { style, type } = props;
    const configImages = {
        // marca
        bch: { source: logoBanco.bch.color, style: Style.logoBch },
        bch_blanco: { source: logoBanco.bch.blanco, style: Style.logoBch },
        bec: { source: logoBanco.bec.color, style: Style.logoBec },
        bec_blanco: { source: logoBanco.bec.blanco, style: Style.logoBec },
        bec_isotipo_blanco: {
            source: logoBanco.bec.isotipo.blanco,
            style: Style.logoCch,
        },
        cch: { source: logoBanco.cch.color, style: Style.logoCch },
        cch_blanco: { source: logoBanco.cch.blanco, style: Style.logoCch },
        bch_isotipo_blanco: {
            source: logoBanco.bch.isotipo.blanco,
            style: Style.logoCch,
        },
        // APPS MI BANCO
        appMiBanco: {
            source: logoApps.miBanco[currentBrandName].color,
            style: Style.logoApp,
        },
        appMiBanco_bch: {
            source: logoApps.miBanco.bch.color,
            style: Style.logoApp,
        },
        appMiBanco_blanco: {
            source: logoApps.miBanco[currentBrandName].blanco,
            style: Style.logoApp,
        },
        appMiBanco_bch_blanco: {
            source: logoApps.miBanco.bch.blanco,
            style: Style.logoApp,
        },
        appMiBanco_bec: {
            source: logoApps.miBanco.bec.color,
            style: Style.logoApp,
        },
        appMiBanco_bec_blanco: {
            source: logoApps.miBanco.bec.blanco,
            style: Style.logoApp,
        },
        appMiBanco_cch: {
            source: logoApps.miBanco.cch.color,
            style: Style.logoApp,
        },
        appMiBanco_cch_blanco: {
            source: logoApps.miBanco.cch.blanco,
            style: Style.logoApp,
        },
        // APPS MI PAGO
        appMiPago: {
            source: logoApps.miPago[currentBrandName].color,
            style: Style.logoAppMiPago,
        },
        appMiPago_bch: {
            source: logoApps.miPago.bch.color,
            style: Style.logoAppMiPago,
        },
        appMiPago_blanco: {
            source: logoApps.miPago[currentBrandName].blanco,
            style: Style.logoAppMiPago,
        },
        appMiPago_bch_blanco: {
            source: logoApps.miPago.bch.blanco,
            style: Style.logoAppMiPago,
        },
        appMiPago_bec: {
            source: logoApps.miPago.bec.color,
            style: Style.logoAppMiPago,
        },
        appMiPago_bec_blanco: {
            source: logoApps.miPago.bec.blanco,
            style: Style.logoAppMiPago,
        },
        appMiPago_cch: {
            source: logoApps.miPago.cch.color,
            style: Style.logoAppMiPago,
        },
        appMiPago_cch_blanco: {
            source: logoApps.miPago.cch.blanco,
            style: Style.logoAppMiPago,
        },
        // APPS MI SEGURO
        appMiSeguro: {
            source: logoApps.miSeguro[currentBrandName].color,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_bch: {
            source: logoApps.miSeguro.bch.color,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_blanco: {
            source: logoApps.miSeguro[currentBrandName].blanco,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_bch_blanco: {
            source: logoApps.miSeguro.bch.blanco,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_bec: {
            source: logoApps.miSeguro.bec.color,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_bec_blanco: {
            source: logoApps.miSeguro.bec.blanco,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_cch: {
            source: logoApps.miSeguro.cch.color,
            style: Style.logoAppMiSeguro,
        },
        appMiSeguro_cch_blanco: {
            source: logoApps.miSeguro.cch.blanco,
            style: Style.logoAppMiSeguro,
        },
        // APPS MI INVERSION
        appMiInversion: {
            source: logoApps.miInversion[currentBrandName].color,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_bch: {
            source: logoApps.miInversion.bch.color,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_blanco: {
            source: logoApps.miInversion[currentBrandName].blanco,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_bch_blanco: {
            source: logoApps.miInversion.bch.blanco,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_bec: {
            source: logoApps.miInversion.bec.color,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_bec_blanco: {
            source: logoApps.miInversion.bec.blanco,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_cch: {
            source: logoApps.miInversion.cch.color,
            style: Style.logoAppMiInversion,
        },
        appMiInversion_cch_blanco: {
            source: logoApps.miInversion.cch.blanco,
            style: Style.logoAppMiInversion,
        },
        // APPS MI PASS
        appMiPass: {
            source: logoApps.miPass[currentBrandName].color,
            style: Style.logoAppMiPass,
        },
        appMiPass_bch: {
            source: logoApps.miPass.bch.color,
            style: Style.logoAppMiPass,
        },
        appMiPass_blanco: {
            source: logoApps.miPass[currentBrandName].blanco,
            style: Style.logoAppMiPass,
        },
        appMiPass_bch_blanco: {
            source: logoApps.miPass.bch.blanco,
            style: Style.logoAppMiPass,
        },
        appMiPass_bec: {
            source: logoApps.miPass.bec.color,
            style: Style.logoAppMiPass,
        },
        appMiPass_bec_blanco: {
            source: logoApps.miPass.bec.blanco,
            style: Style.logoAppMiPass,
        },
        appMiPass_cch: {
            source: logoApps.miPass.cch.color,
            style: Style.logoAppMiPass,
        },
        appMiPass_cch_blanco: {
            source: logoApps.miPass.cch.blanco,
            style: Style.logoAppMiPass,
        },
        // APPS BANCONEXION
        appBanconexion: {
            source: logoApps.banconexion.color,
            style: Style.logoBanconexion,
        },
        appBanconexion_blanco: {
            source: logoApps.banconexion.blanco,
            style: Style.logoBanconexion,
        },
        // COMBINADOS MI BANCO
        combined_miBanco_bch_blanco: {
            source: logoCombined.miBanco.bch.blanco,
            style: Style.logoCombinedMiPagoBCH,
        },
        combined_miBanco_bec_blanco: {
            source: logoCombined.miBanco.bec.blanco,
            style: Style.logoCombinedMiPagoBEC,
        },
        combined_miBanco_cch_blanco: {
            source: logoCombined.miBanco.cch.blanco,
            style: Style.logoCombinedMiPagoCCH,
        },
        // COMBINADOS MI PAGO
        combined_miPago_bch_blanco: {
            source: logoCombined.miPago.bch.blanco,
            style: Style.logoCombinedMiPagoBCH,
        },
        combined_miPago_bec_blanco: {
            source: logoCombined.miPago.bec.blanco,
            style: Style.logoCombinedMiPagoBEC,
        },
        combined_miPago_cch_blanco: {
            source: logoCombined.miPago.cch.blanco,
            style: Style.logoCombinedMiPagoCCH,
        },
        // COMBINADOS MI SEGURO
        combined_miSeguro_bch_blanco: {
            source: logoCombined.miSeguro.bch.blanco,
            style: Style.logoCombinedMiSeguroBCH,
        },
        combined_miSeguro_bec_blanco: {
            source: logoCombined.miSeguro.bec.blanco,
            style: Style.logoCombinedMiSeguroBEC,
        },
        combined_miSeguro_cch_blanco: {
            source: logoCombined.miSeguro.cch.blanco,
            style: Style.logoCombinedMiSeguroCCH,
        },
        // COMBINADOS MI INVERSION
        combined_miInversion_bch_blanco: {
            source: logoCombined.miInversion.bch.blanco,
            style: Style.logoCombinedMiInversionBCH,
        },
        combined_miInversion_bec_blanco: {
            source: logoCombined.miInversion.bec.blanco,
            style: Style.logoCombinedMiInversionBEC,
        },
        combined_miInversion_cch_blanco: {
            source: logoCombined.miInversion.cch.blanco,
            style: Style.logoCombinedMiInversionCCH,
        },
        // COMBINADOS MI PASS
        // combined_miPass_bch_blanco: { source: logoCombined.miPass.bch.blanco, style: Style.logoCombinedMiPassBCH },
        // combined_miPass_bec_blanco: { source: logoCombined.miPass.bec.blanco, style: Style.logoCombinedMiPassBEC },
        // combined_miPass_cch_blanco: { source: logoCombined.miPass.cch.blanco, style: Style.logoCombinedMiPassCCH },
        // COMBINADOS BANCONEXION
        combined_banconexion_bch_blanco: {
            source: logoCombined.banconexion.bch.blanco,
            style: Style.logoCombinedMiPagoBCH,
        },
        // ISOTIPOS
        isotipo_miBanco: {
            source: isotipos.miBanco.blanco,
            style: Style.isotipo,
        },
        isotipo_miSeguro: {
            source: isotipos.miSeguro.blanco,
            style: Style.isotipo,
        },
        isotipo_miInversion: {
            source: isotipos.miInversion.blanco,
            style: Style.isotipo,
        },
        isotipo_miPass: {
            source: isotipos.miPass.blanco,
            style: Style.isotipo,
        },
        isotipo_miPago: {
            source: isotipos.miPago.blanco,
            style: Style.isotipo,
        },
    };
    const imageProps = configImages[type];
    return (
        <Image
            {...props}
            source={imageProps?.source}
            style={[imageProps?.style, style]}
        />
    );
};

Logo.propTypes = {
    /**
     * Estilos.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Nombre de la app o marca.
     */
    type: PropTypes.string,
};
Logo.defaultProps = {
    style: {},
    type: 'bch',
};

export default memo(Logo);
