// react-native-minerva/src/components/logo/style.js
import { StyleSheet } from 'react-native';
import { Base, Spacing, Scale } from '../../index';

const style = StyleSheet.create({
    logoBch: {
        width: Scale(98),
        height: Base,
        alignSelf: 'center',
        resizeMode: 'contain',
    },
    logoBec: {
        width: Scale(98),
        height: Spacing.spacer_2,
        alignSelf: 'center',
        resizeMode: 'contain',
    },
    logoCch: {
        width: Scale(102),
        height: Spacing.spacer_3,
        alignSelf: 'center',
        resizeMode: 'contain',
    },
    logoApp: {
        width: Scale(87),
        height: Scale(27),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoAppMiPago: {
        width: Scale(70),
        height: Scale(14),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoAppMiSeguro: {
        width: Scale(87),
        height: Scale(27),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoAppMiInversion: {
        width: Scale(87),
        height: Scale(27),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoAppMiPass: {
        width: Scale(70),
        height: Scale(14),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoBanconexion: {
        width: Scale(104),
        height: Scale(18),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPagoBCH: {
        width: Scale(172),
        height: Scale(60),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPagoBEC: {
        width: Scale(234),
        height: Scale(51),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPagoCCH: {
        width: Scale(204),
        height: Scale(56),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiSeguroBCH: {
        width: Scale(172),
        height: Scale(60),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiSeguroBEC: {
        width: Scale(234),
        height: Scale(51),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiSeguroCCH: {
        width: Scale(204),
        height: Scale(56),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiInversionBCH: {
        width: Scale(172),
        height: Scale(60),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiInversionBEC: {
        width: Scale(234),
        height: Scale(51),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiInversionCCH: {
        width: Scale(204),
        height: Scale(56),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPassBCH: {
        width: Scale(172),
        height: Scale(60),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPassBEC: {
        width: Scale(234),
        height: Scale(51),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    logoCombinedMiPassCCH: {
        width: Scale(204),
        height: Scale(56),
        alignSelf: 'center',
        resizeMode: 'contain',
        marginBottom: Base,
    },
    isotipo: {
        width: Scale(80),
        height: Scale(100),
        resizeMode: 'contain',
        marginBottom: Base,
    },
});

export default style;
