// react-native-minerva/scripts/generate/items.js
const filename = require('file-name');
const replace = require('replace-in-file');

const items = [
    {
        option: 'Crear componente',
        defaultCase: '(pascalCase)',
        entry: {
            folderPath: './scripts/generate/templates/component',
        },
        stringReplacers: ['__nombreComponente__'],
        output: {
            path: './src/components/__nombreComponente__(camelCase)',
            pathAndFileNameDefaultCase: '(pascalCase)',
            overwrite: false,
        },
        onComplete: async (results) => {
            console.log(`results`, results);
            await importVuexStore(results);
        },
    },
    {
        option: 'Crear constante',
        defaultCase: '(pascalCase)',
        entry: {
            folderPath: './scripts/generate/templates/constant',
        },
        stringReplacers: ['__nombreConstante__'],
        output: {
            path: './src/constants',
            pathAndFileNameDefaultCase: '(pascalCase)',
            overwrite: true,
        },
        onComplete: async (results) => {
            console.log(`results`, results);
            await importVuexStore(results, 'constants');
        },
    },
];

async function importVuexStore(results, recurso = 'component') {
    const replace = require('replace-in-file');
    const file = results.output.files[0];
    const path = file.replace('src/', '').split('.').slice(0, -1).join('.');
    const componentName = filename(path);
    const importComponent = `import ${componentName} from '${path}';`;
    const pathFileReplace = './src/index.js';
    const options = {
        export: {
            files: pathFileReplace,
            from: '//ADD_EXPORT',
            to: `${componentName},
    //ADD_EXPORT`,
        },
        import_constast: {
            files: pathFileReplace,
            from: '//ADD_IMPORT_CONSTANT',
            to: `${importComponent}
//ADD_IMPORT_CONSTANT`,
        },
        import_component: {
            files: pathFileReplace,
            from: '//ADD_IMPORT_COMPONENT',
            to: `${importComponent}
//ADD_IMPORT_COMPONENT`,
        }
    };

    if (recurso === 'component') {
        try {
            const results_export = await replace(options.export);
            const results_import_component = await replace(options.import_component);
            console.log('results_export:', results_export);
            console.log('results_import_component:', results_import_component);
        }
        catch (error) {
            console.error('Error occurred:', error);
        }
    } else {
        try {
            const results_export = await replace(options.export);
            const results_import_constant = await replace(options.import_constast);
            console.log('results_export:', results_export);
            console.log('resuresults_import_constantlts:', results_import_constant);
        }
        catch (error) {
            console.error('Error occurred:', error);
        }
    }
}

exports.items = items;
