// react-native-minerva/src/components/cardItemProducto/components/CardHeader/CardState.js
import React, { memo } from 'react';
import { Badge } from '../../../../index';
import CARDITEMPRODUCTO_STR from '../../CardItemProductoStr';
import { tarjetaEstadoTypes } from '../../type';

/**
 * Componente para mostrar estado del producto en el area del derecha del encabezado del producto.
 *
 * @component
 */
const CardState = ({ tarjetaEstado }) => {
    const getBadgePropsByState = () => {
        const iconLock = 'icon-bch-lock';
        const configBadge = {};
        configBadge[CARDITEMPRODUCTO_STR.estado.activa] = {
            success: true,
            text: CARDITEMPRODUCTO_STR.estado.activa,
        };
        configBadge[CARDITEMPRODUCTO_STR.estado.permanente] = {
            danger: true,
            text: CARDITEMPRODUCTO_STR.estado.permanente,
            icon: iconLock,
        };
        configBadge[CARDITEMPRODUCTO_STR.estado.temporal] = {
            info: true,
            text: CARDITEMPRODUCTO_STR.estado.temporal,
            icon: iconLock,
        };
        configBadge[CARDITEMPRODUCTO_STR.estado.contratado] = {
            info: true,
            text: CARDITEMPRODUCTO_STR.estado.contratado,
        };
        configBadge[CARDITEMPRODUCTO_STR.estado.bloqueado] = {
            danger: true,
            text: CARDITEMPRODUCTO_STR.estado.bloqueado,
            icon: iconLock,
        };
        return configBadge[tarjetaEstado.toUpperCase()];
    };

    const badgeProps = getBadgePropsByState();

    return (
        <Badge
            success={badgeProps?.success}
            danger={badgeProps?.danger}
            info={badgeProps?.info}
            icon={badgeProps?.icon}
            text={badgeProps?.text}
        />
    );
};

CardState.propTypes = {
    tarjetaEstado: tarjetaEstadoTypes,
};
CardState.defaultProps = {
    tarjetaEstado: CARDITEMPRODUCTO_STR.estado.bloqueado,
};

export default memo(CardState);
