// react-native-minerva/src/components/cardItemProducto/components/CardHeader/LogoMarcaTarjeta.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Image } from 'react-native';
import Images from 'bch-assets';
import { PRODUCTS } from 'bch-mobile-constants';
import { HelperStyle, Icon, Validate } from '../../../../index';
import Style from '../../style';
import CARDITEMPRODUCTO_STR from '../../CardItemProductoStr';
import { isTextWhite } from '../../utils';

/*
 * Renderiza el logo Visa (formato: icono) o Mastercard (formato: imagen).
 * Obtiene icono tarjeta: Visa o MasterCard.
 */
const LogoMarcaTarjeta = ({ icon, tipoTarjeta, marca }) => {
    const {
        components: {
            cardItemProducto: { mastercard },
        },
    } = Images;
    const {
        BRAND: { VISA },
    } = PRODUCTS;
    /**
     * Obtiene icono tarjeta: Visa o MasterCard.
     */
    const renderIconVisaOMastercard = () => (
        <Icon
            name={icon}
            style={[
                Style.icon,
                HelperStyle.ai_fs,
                HelperStyle.jc_c,
                HelperStyle.fs_6,
                HelperStyle.lh_3,
                HelperStyle.gray_dark,
                Validate.style(isTextWhite(tipoTarjeta), HelperStyle.white),
            ]}
            accessible
            accessibilityRole="image"
            accessibilityLabel={CARDITEMPRODUCTO_STR.logoTrajeta}
            testID={CARDITEMPRODUCTO_STR.logoTrajeta}
        />
    );

    /*
     * Renderiza el logo Visa (formato: icono) o Mastercard (formato: imagen).
     */
    const renderMarca = () => {
        if (marca === VISA) {
            return (
                <Icon
                    name="icon-bch-visa-brand"
                    style={[
                        Style.icon,
                        HelperStyle.ai_fs,
                        HelperStyle.jc_c,
                        HelperStyle.fs_3,
                        HelperStyle.lh_3,
                        HelperStyle.brand,
                        Validate.style(
                            isTextWhite(tipoTarjeta),
                            HelperStyle.white,
                        ),
                    ]}
                    accessible
                    accessibilityRole="image"
                    accessibilityLabel={CARDITEMPRODUCTO_STR.logoTrajetaVisa}
                    testID={CARDITEMPRODUCTO_STR.logoTrajetaVisa}
                />
            );
        }
        return (
            <Image
                source={mastercard}
                style={Style.styleLogoMastercard}
                accessible
                accessibilityRole="image"
                accessibilityLabel={CARDITEMPRODUCTO_STR.logoTrajetaMastercard}
                testID={CARDITEMPRODUCTO_STR.logoTrajetaMastercard}
            />
        );
    };
    return icon ? renderIconVisaOMastercard() : renderMarca();
};

LogoMarcaTarjeta.propTypes = {
    icon: PropTypes.string,
    tipoTarjeta: PropTypes.string,
    marca: PropTypes.string,
};
LogoMarcaTarjeta.defaultProps = {
    icon: '',
    tipoTarjeta: '',
    marca: '',
};

export default memo(LogoMarcaTarjeta);
