// bch-mobile-compartir-cuenta/src/components/CompartirCuenta.js
import React, { memo } from 'react';
import Share from 'react-native-share';
import PropTypes from 'prop-types';
import {
    Button,
    HelperStyle,
    Text,
} from 'react-native-minerva';
import labels from '../labelAutomatizacion';

const propTypes = {
    title: PropTypes.string,
    data: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.bool,
        PropTypes.string,
        PropTypes.array,
    ]),
    style: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.bool,
    ]),
    colorIcon: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.bool,
    ]),
    onPress: PropTypes.func,
};

const defaultProps = {
    title: 'Compartir cuenta',
    data: false,
    style: false,
    colorIcon: HelperStyle.white,
    onPress: () => {} 
};

const CompartirCuenta = ({ title, data, style, colorIcon, onPress }) => {
    /**
     * Realiza compartir nativo de datos de la cuenta bancaria.
     * Se formatea algunos datos y se levanta el compartir nativo para iOS y Android.
     */
    const shareNative = () => {
        /*
        * Creamos la estructura data solicitada para el componente Share.
        */
        const shareDataText = {
            title: title,
            url: null,
            type: 'text',
            message: data,
        };
        onPress()
        return Share.open(shareDataText);
    };

    return data && (
        <Button
            inverse
            smaller
            style={style}
            onPress={shareNative}
            accessibilityLabel={labels.compartir.BOTON_COMPARTIR_CUENTA}
            testID={labels.compartir.BOTON_COMPARTIR_CUENTA}
        >
            <Text smaller bold style={[HelperStyle.pa_0, HelperStyle.ma_0, colorIcon]}>{labels.compartir.TEXTO_COMPARTIR_CUENTA}</Text>
        </Button>
    );
};

CompartirCuenta.propTypes = propTypes;
CompartirCuenta.defaultProps = defaultProps;

export default memo(CompartirCuenta);
