// react-native-minerva/src/components/input/components/style.js
import { StyleSheet } from 'react-native';
import { minerva, Spacing, Scale, ColorsBCH } from '../../../index';

const { inputHeightBase } = minerva();

const Style = (props) => {
    const { colorBorder, inputHeight } = props;
    const borderColor = colorBorder();
    return StyleSheet.create({
        inputHeight: {
            height: inputHeight,
        },
        borderInputRight: {
            borderTopRightRadius: Spacing.spacer_2,
            borderBottomRightRadius: Spacing.spacer_2,
            flex: 1,
        },
        inputBorderColor: {
            borderColor,
        },
        inputHeightBase: {
            height: inputHeightBase,
        },
        inputHeightBaseMultiline: {
            minHeight: inputHeightBase,
        },
        borderInputContainer: {
            width: '100%',
            position: 'absolute',
            flexDirection: 'row',
        },
        borderTopTransparent: {
            borderTopColor: ColorsBCH.transparent,
        },
        borderInputLeft: {
            borderTopLeftRadius: Spacing.spacer_2,
            borderBottomLeftRadius: Spacing.spacer_2,
            width: Scale(11),
        },
    });
};

export default Style;
