// react-native-minerva/src/constants/spacing.js
import _ from 'lodash';
import { Platform, Dimensions } from 'react-native';

const { height } = Dimensions.get('window');

export const defaultBase = 16;
export const smallBase = 14;

const getBase = () => {
    if (Platform.OS === 'ios' && height <= 568) {
        return smallBase;
    }
    return defaultBase;
};

const Base = getBase();

export const Scale = (size) => {
    const calculateScale = size / defaultBase;
    if (Platform.OS === 'ios' && height <= 568) {
        return smallBase * calculateScale;
    }
    return size;
};

const HelperSpacing = {
    spacer: {
        _0: 0 * Base, // 0
        _1: 0.25 * Base, // 4
        _2: 0.5 * Base, // 8
        _3: 0.75 * Base, // 12
        _4: 1 * Base, // 16
        _5: 1.5 * Base, // 24
        _6: 2 * Base, // 32
        _7: 2.5 * Base, // 40
    },
    type: {
        _1: 0.75 * Base, // 12
        _2: 0.875 * Base, // 14
        _3: 1 * Base, // 16
        _4: 1.25 * Base, // 20
        _5: 1.5 * Base, // 24
        _6: 1.625 * Base, // 26
        _7: 2 * Base, // 32
        _8: 2.5 * Base, // 40
    },
    radius: {
        _0: 0 * Base,
        _1: 0.25 * Base, // 4
        _2: 0.5 * Base, // 8
        _3: 1 * Base, // 16
    },
    lineHeight: {
        _1: 1 * Base, // 16
        _2: 1.5 * Base, // 24
        _3: 2 * Base, // 32
    },
    border: {
        _0: 0,
        _1: 1,
        _2: 2,
        _4: 4,
    },
};

const joinSpacing = (object) => {
    const scaled = {};
    _.forEach(object, (v, i) => {
        _.forEach(v, (vv, ii) => {
            scaled[`${i}${ii}`] = vv;
        });
    });
    return scaled;
};

const Spacing = joinSpacing(HelperSpacing);

export { HelperSpacing, Spacing, Base };
