// bch-mobile-menu/src/components/AvatarName.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Avatar, View, HelperStyle } from 'react-native-minerva';
import Utils from 'bch-mobile-utils';

const propTypes = {
    accessibilityLabel: PropTypes.string,
    image: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    lastName: PropTypes.string,
    loading: PropTypes.bool,
    name: PropTypes.string,
    testID: PropTypes.string,
};

const defaultProps = {
    accessibilityLabel: '',
    image: false,
    lastName: '',
    loading: false,
    name: '',
    testID: '',
};

const AvatarName = ({ name, lastName, loading, accessibilityLabel, testID, image }) => (
    <View style={HelperStyle.mr_3}>
        <Avatar
            hasBorderAvatarImage
            label={Utils.getInitials(name, lastName)}
            name={name}
            lastName={lastName}
            loading={loading}
            accessibilityLabel={accessibilityLabel}
            testID={testID}
            image={image}
        />
    </View>
);

AvatarName.propTypes = propTypes;
AvatarName.defaultProps = defaultProps;

export default memo(AvatarName);
