// bch-mobile-constants/src/bines.js
import Products from './products';
const {
    CARD_TYPE: {
        DEFAULT,
        GOLD,
        SILVER,
        BLACK,
    },
} = Products;
/**
* Valores de bines
*/
const BINES_VALUES = {
    DEFAULT: DEFAULT,
    GOLD: GOLD,
    SILVER: SILVER,
    BLACK: BLACK,
    BCOBCH: 'BCH',
    BCOBEC: 'BEC',
    BCOCCH: 'CCH',
};
/**
* Json Bines
*/
const BINES = [
    { bin: '415282', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '415282', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '415283', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.GOLD },
    { bin: '415283', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.GOLD },
    { bin: '422766', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '422766', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '422777', bco: BINES_VALUES.BCOCCH, color: BINES_VALUES.SILVER },
    { bin: '434956', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '434956', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '446867', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.SILVER },
    { bin: '446867', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.SILVER },
    { bin: '448480', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '448480', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '450878', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '450878', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '454413', bco: BINES_VALUES.BCOCCH, color: BINES_VALUES.SILVER },
    { bin: '456865', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '456865', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '459311', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '459311', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '469772', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '469772', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '475774', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '475774', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '475775', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.SILVER },
    { bin: '475775', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.SILVER },
    { bin: '475776', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '475776', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '475777', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '475777', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '476025', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '476025', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '477041', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '477041', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '480293', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '480293', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '480916', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '480916', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '480917', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '480917', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '492210', bco: BINES_VALUES.BCOCCH, color: BINES_VALUES.SILVER },
    { bin: '496670', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.GOLD },
    { bin: '496670', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.GOLD },
    { bin: '498811', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '498811', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '518370', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '518370', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '522832', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.BLACK },
    { bin: '522832', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.BLACK },
    { bin: '534129', bco: BINES_VALUES.BCOCCH, color: BINES_VALUES.GOLD },
    { bin: '540476', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.GOLD },
    { bin: '540476', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.GOLD },
    { bin: '540782', bco: BINES_VALUES.BCOCCH, color: BINES_VALUES.GOLD },
    { bin: '543564', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '543564', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '546689', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.SILVER },
    { bin: '546689', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.SILVER },
    { bin: '548652', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '548652', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.DEFAULT },
    { bin: '548653', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.GOLD },
    { bin: '548653', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.GOLD },
    { bin: '549163', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.SILVER },
    { bin: '549163', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.SILVER },
    { bin: '558685', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '558685', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.GOLD },
    { bin: '558840', bco: BINES_VALUES.BCOBCH, color: BINES_VALUES.DEFAULT },
    { bin: '558840', bco: BINES_VALUES.BCOBEC, color: BINES_VALUES.SILVER }
];

export default { BINES, BINES_VALUES };
