// bch-mobile-constants/src/configApp.js
import { Platform } from 'react-native';
import APP_HEADERS from './headers';
import APP_SECRET from './secret';

const BUNDLE_MIBANCO_IOS = ['cl.bancochile.mibanco3', 'cl.bancochile.mibancoedwards3', 'cl.bancochile.mibancocredichile3', 'cl.bch.osiris'];
const BUNDLE_MIBANCO_ANDROID = ['cl.bancochile.mi_banco', 'cl.bancochile.mi_edwards', 'cl.bancochile.mi_credichile'];

const BUNDLE_MIPAGO_IOS = ['cl.bancochile.mipago'];
const BUNDLE_MIPAGO_ANDROID =  ['cl.bancochile.mi_pago'];

const BUNDLE_MIINVERSION_IOS = ['cl.bancochile.minversion2'];
const BUNDLE_MIINVERSION_ANDROID =  ['cl.bancochile.mi_inversion2'];

const BUNDLE_MISEGURO_IOS = ['cl.bancochile.miseguro'];
const BUNDLE_MISEGURO_ANDROID =  ['cl.bancochile.mi_seguro'];

export default {
    MIBANCO: {
        NOMBRE: 'Mi Banco',
        LOGO: 'appMiBanco',
        PASOS_CUSTOMIZACION: 2,
        BIOMETRIA: true,
        BUNDLE_ANDROID: BUNDLE_MIBANCO_ANDROID,
        BUNDLE_IOS: BUNDLE_MIBANCO_IOS,
        BUNDLEID: Platform.OS === 'ios' ? BUNDLE_MIBANCO_IOS : BUNDLE_MIBANCO_ANDROID,
        URL_SERVICE_APP: '/v1/mi-banco',
        CARRUSEL_PRODUCTOS_MOSTRAR_LINEAS_CREDITO: true,
        ORIGEN_TRANSACTION_TYPE: 'MIBANCO',
        FIREBASE: true,
        HEADERS: APP_HEADERS,
    },
    MIPAGO: {
        NOMBRE: 'Mi Pago',
        LOGO: 'appMiPago',
        PASOS_CUSTOMIZACION: 4,
        BIOMETRIA: false,
        BUNDLE_ANDROID: BUNDLE_MIPAGO_ANDROID,
        BUNDLE_IOS: BUNDLE_MIPAGO_IOS,
        BUNDLEID: Platform.OS === 'ios' ? BUNDLE_MIPAGO_IOS : BUNDLE_MIPAGO_ANDROID,
        URL_SERVICE_APP: '/v1/mi-pago',
        CARRUSEL_PRODUCTOS_MOSTRAR_LINEAS_CREDITO: true,
        ORIGEN_TRANSACTION_TYPE: '',
        FIREBASE: true,
        HEADERS: APP_HEADERS,
    },
    MIINVERSION: {
        NOMBRE: 'Mi Inversión',
        LOGO: 'appMiInversion',
        PASOS_CUSTOMIZACION: 1,
        BIOMETRIA: false,
        BUNDLE_ANDROID: BUNDLE_MIINVERSION_ANDROID,
        BUNDLE_IOS: BUNDLE_MIINVERSION_IOS,
        BUNDLEID: Platform.OS === 'ios' ? BUNDLE_MIINVERSION_IOS : BUNDLE_MIINVERSION_ANDROID,
        URL_SERVICE_APP: '',
        CARRUSEL_PRODUCTOS_MOSTRAR_LINEAS_CREDITO: false,
        SERVICES: {
            PRODUCTOS: {
                GET_URL_PRODUCTOS: '/clientes',
                GET_URL_SALDO_CUENTA_CORRIENTE: '/productos/cuentas-corrientes',
            },
            DOLARES: {
                POST_URL_CAJA_VALIDACION: '/productos/divisas/caja/validacion',
                POST_URL_CAJA: '/productos/divisas/caja',
                GET_URL_DOLAR_COMPRA_VENTA: '/utilidades/indicadores/dolar/compra-venta',
                POST_URL_ONLINE_VALIDACION: '/productos/divisas/online/validacion',
                POST_URL_ONLINE: '/productos/divisas/online',
                GET_URL_PARAMETROS: '/productos/divisas/enrolamiento/parametros/',
            },
        },
        ORIGEN_TRANSACTION_TYPE: 'MIINVERSION',
        FIREBASE: true,
        HEADERS: { ...APP_HEADERS, ...APP_SECRET },
    },
    MISEGURO: {
        NOMBRE: 'Mi Seguro',
        LOGO: 'appMiSeguro',
        BUNDLE_ANDROID: BUNDLE_MISEGURO_ANDROID,
        BUNDLE_IOS: BUNDLE_MISEGURO_IOS,
        BUNDLEID: Platform.OS === 'ios' ? BUNDLE_MISEGURO_IOS : BUNDLE_MISEGURO_ANDROID,
        URL_SERVICE_APP: '/v1/mi-seguro',
        SERVICES: {
            SERVICIOS: '/servicios',
            IMAGENES: '/imagenes',
            CLIENTES: '/clientes',
        },
        PASOS_CUSTOMIZACION: 0,
        BIOMETRIA: false,
        HEADERS: APP_HEADERS,
    },
    ENROLAMIENTO: {
        PASOS: 3,
        PASOS_BIOMETRIA: 4
    }
};
