// bch-mobile-constants/src/products.js
export default {
    CODE_PRODUCT: {
        CTD: 'CTD',
        TNM: 'TNM',
        RTV: 'RTV',
        LCD: 'LCD',
        JUV: 'JUV',
        CEX: 'CEX',
        AHO: 'AHO',
        TMB: 'TMB',
    },
    NAME_PRODUCT: {
        CUENTA_CORRIENTE: 'Cuenta Corriente',
        CUENTA_CORRIENTE_ME: 'Cuenta Corriente M/E',
        TARJETA_CREDITO: 'Tarjeta de Crédito',
        LINEA_CREDITO: 'Línea de Crédito',
        CUENTA_VISTA: 'Cuenta Vista',
        CUENTA_AHORRO: 'Cuenta de Ahorro',
        CUENTA_FAN: 'Cuenta FAN',
        CREDITO: 'Crédito',
        DEBITO: 'Débito',
        CUENTA_VISTA_EMPRESA: 'Cuenta Vista Empresas',
        DAP: 'Depósito a Plazo',
        CUENTA_AHORRO_ICONO: 'Cuenta de Ahorro',
        CAJA: 'Mi caja Banchile',
        FFMM: 'Fondos Mutuos',
        ACCIONES_NACIONALES: 'Acciones Nacionales',
        ACCIONES_INTERNACIONALES: 'Acciones Internacionales',
        ETF: 'ETF iShares',
        COMPRA_DOLARES: 'Compra de Dólares',
        COMPRA_VENTA_DOLARES: 'Compra o vende Dólares',
        FONDOS_INVERSION: 'Fondos de Inversión',
        APV: 'APV',
        OTROS: 'Otros Productos',
        PERSHING: 'Inversiones en Pershing',
    },
    CLASS_ACCOUNT: {
        CVISTA: 'CVISTA',
        CVICCH: 'CVICCH',
        VTACNN: 'VTACNN',
        FAN: 'FAN',
        CVIEMP: 'CVIEMP',
        CVICH: 'CVICH',
    },
    PRODUCT_TYPE: {
        PRODUCT_TC: 'TARJETA_CREDITO',
        PRODUCT_LC: 'LINEA_CREDITO',
        PRODUCT_CC: 'CUENTA_CORRIENTE',
        PRODUCT_CC_ME: 'CUENTA_CORRIENTE_ME',
        PRODUCT_CV: 'CUENTA_VISTA',
        PRODUCT_LC_PAY: 'LINEA_CREDITO_PAGO',
        PRODUCT_CF: 'CUENTA_FAN',
        PRODUCT_CA: 'CUENTA_AHORRO',
        PRODUCT_CVE: 'CUENTA_VISTA_EMPRESA',
        PRODUCT_CA_MI_INVERSION: 'CUENTA_AHORRO_ICONO',
        PRODUCT_DAP: 'DAP',
        PRODUCT_BILLETERA: 'BILLETERA',
        PRODUCT_FFMM: 'FFMM',
        PRODUCT_ACCIONES_NACIONALES: 'ACCIONES_NACIONALES',
        PRODUCT_ACCIONES_INTERNACIONALES: 'ACCIONES_INTERNACIONALES',
        PRODUCT_ETF: 'ETF',
        PRODUCT_COMPRA_DOLARES: 'COMPRA_DOLARES',
        PRODUCT_COMPRA_VENTA_DOLARES: 'COMPRA_VENTA_DOLARES',
        PRODUCT_FONDOS_INVERSION: 'FONDOS_INVERSION',
        PRODUCT_APV: 'APV',
        PRODUCT_OTROS: 'OTROS_PRODUCTOS',
        PRODUCT_PERSHING: 'PERSHING',
        OTROS_BANCOS: 'OTROS_BANCOS',
    },
    PRODUCT_DETAIL: {
        DISPONIBLE: 'Disponible',
        UTILIZADO: 'Utilizado',
        CUPO_TOTAL: 'Cupo total',
        NACIONAL: 'NACIONAL',
        INTERNACIONAL: 'INTERNACIONAL',
    },
    CARD_TYPE: {
        BLACK: 'BLACK',
        BRAND: 'BRAND',
        DEFAULT_BCH: 'DEFAULT_BCH',
        DEFAULT_BEC: 'DEFAULT_BEC',
        DEFAULT_CCH: 'DEFAULT_CCH',
        DEFAULT: 'DEFAULT',
        GOLD: 'GOLD',
        GRAY: 'GRAY',
        SILVER: 'SILVER',
        WHITE: 'WHITE',
    },
    BRAND: {
        VISA: 'VISA',
        MASTERCARD: 'MASTERCARD',
    },
    SHORT_NAME_ACCOUNT: {
        CUENTA_CORRIENTE: 'C. Corriente',
        CUENTA_CORRIENTE_ME: 'C. Corriente M/E',
        CUENTA_VISTA: 'C. Vista',
        CUENTA_AHORRO: 'C. Ahorro',
        CUENTA_VISTA_EMPRESA: 'C. Vista Empresa',
    },
    OWNERSHIP_TC: {
        TITULAR: 'T',
        ADICIONAL: 'A',
    },
    CURRENCY_TYPE: {
        CLP: 'CLP',
        USD: 'USD',
    },
    ERROR: {
        ERROR_CUENTA: 'No hemos podido cargar los datos de tu cuenta',
        ERROR_CUENTA_SALDO: 'No es posible cargar los saldos de la cuenta',
        ERROR_TARJETA: 'No hemos podido cargar los datos de tu tarjeta',
        ERROR_TARJETA_SALDO: 'No es posible cargar los saldos de la tarjeta',
        BTN_REINTENTAR: 'Reintentar',
    }
};
