// bch-mobile-constants/src/transactionsMiPass.js
const TRANSACTIONS = {
    TEF: 'MIBANCO_TEF_2',
    TEF_TC: 'MIBANCO_TEFTC',
    CREATE_RECIPIENT: 'MIBANCO_CREADEST',
    EDIT_RECIPIENT: 'MIBANCO_EDITARDESTINATARIO',
    ADD_BILL: 'MB_CREACUENTA',
    CUENTA_FAN: 'MIBANCO_TDCVV',
    AVANCE_TC: 'MIBANCO_AVANCE_TC',
    CAMBIO_PINPASS: 'MB_SeteoPinPass',
    DOLAR_CAJA: 'MIBANCO_CDIVCAJA',
    DOLAR_VENTA: 'MIBANCO_VDIVISAS',
    DOLAR_COMPRA: 'MIBANCO_CDIVISAS',
    MIINVERSION: {
        DOLAR_CAJA: 'MIINVERSION_CDVCAJA',
        DOLAR_VENTA: 'MIINVERSION_VDIVISAS',
        DOLAR_COMPRA: 'MIINVERSION_CDIVISAS',
    },
    MIBANCO: {
        TEF: 'MIBANCO_TEF_2',
        CREATE_RECIPIENT: 'MIBANCO_CREADEST',
        EDIT_RECIPIENT: 'MIBANCO_EDITARDESTINATARIO',
        ADD_BILL: 'MB_CREACUENTA',
        CUENTA_FAN: 'MIBANCO_TDCVV',
        AVANCE_TC: 'MIBANCO_AVANCE_TC',
        CAMBIO_PINPASS: 'MB_SeteoPinPass',
        DOLAR_CAJA: 'MIBANCO_CDIVCAJA',
        DOLAR_VENTA: 'MIBANCO_VDIVISAS',
        DOLAR_COMPRA: 'MIBANCO_CDIVISAS',
    },
}

export default TRANSACTIONS;
