// react-native-minerva/src/components/modalContextual/type.js
import PropTypes from 'prop-types';
import MODALCONTEXTUAL_STR from './ModalContextualStr';

const propTypes = {
    /**
     * Boleano que indica si mostrar o ocultar el modal.
     */
    visible: PropTypes.bool.isRequired,
    /**
     * Función para cambiar el estado de visible.
     */
    setVisible: PropTypes.oneOfType([PropTypes.func, PropTypes.bool])
        .isRequired,
    /**
     * Objecto que contiene los datos a mostrar en el header: Icono, titulo y subtitulo.
     */
    headerData: PropTypes.shape({
        iconName: PropTypes.string,
        title: PropTypes.string,
        subtitle: PropTypes.string,
    }),
    /**
     * Arreglo que contiene la lista de opciones a mostrar.
     */
    listOptions: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string,
            iconName: PropTypes.string,
            color: PropTypes.string,
            onPress: PropTypes.func,
            testID: PropTypes.string,
            accesibilityLabel: PropTypes.string,
        }),
    ),
    /**
     * (Opcional) Label para el botón en el footer por defecto 'Volver'.
     */
    buttonFooterName: PropTypes.string,
};

const defaultProps = {
    setVisible: false,
    listData: MODALCONTEXTUAL_STR.listOptions,
    buttonFooterName: MODALCONTEXTUAL_STR.btnVolver,
};

export { propTypes, defaultProps };
