// react-native-minerva/src/components/buttonsGroup/types.js
import PropTypes from 'prop-types';

const propTypes = {
    /**
     * Objecto con propiedades para configurar los botones.
     */
    items: PropTypes.arrayOf(PropTypes.object),
    /**
     * Boleano para definir alineación horizontal de los botones.
     */
    horizontal: PropTypes.bool,
    /**
     * Boleano para tamaño de los botones de tipo small.
     */
    small: PropTypes.bool,
    /**
     * Estilos del contenedor de botones.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Estilos del text dentro de los botones.
     */
    styleText: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Estilos para los botones.
     */
    styleItems: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Boleano para habilitar y deshabilitar los botones.
     */
    disabled: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Estilos para definir estado de carga.
     */
    loading: PropTypes.bool,
    /**
     * Boleano para definir que los botones pueden fluir en varias líneas.
     */
    inline: PropTypes.bool,
    /**
     * Boleano para definir botón con ancho al 100%.
     */
    block: PropTypes.bool,
};
const defaultProps = {
    items: [],
    horizontal: false,
    small: false,
    style: {},
    styleText: {},
    styleItems: {},
    disabled: undefined,
    loading: false,
    inline: false,
    block: false,
};

export { propTypes, defaultProps };
