// react-native-minerva/src/components/footer/style.js
import { StyleSheet } from 'react-native';
import { Spacing, dimensions } from '../../index';

const Style = ({ HelperStyle, variables }) => {
    const shadow = {
        ...HelperStyle.shadow_top,
        ...HelperStyle.bg_white,
    };
    const rounded = {
        ...HelperStyle.btrr_3,
        ...HelperStyle.btlr_3,
        ...HelperStyle.bblr_0,
        ...HelperStyle.bbrr_0,
    };
    return StyleSheet.create({
        positionStyle: {
            ...HelperStyle.p_a,
            ...HelperStyle.b_0,
        },
        footer: {
            paddingTop: Spacing.spacer_4,
            paddingHorizontal: Spacing.spacer_4,
            minHeight: variables.footerHeight,
            paddingBottom: variables.footerPaddingBottom,
            elevation: 0,
            width: dimensions.fullWidth,
        },
        tabBar: {
            flexDirection: 'row',
            justifyContent: 'center',
            paddingTop: 0,
            paddingHorizontal: Spacing.spacer_4,
            paddingBottom: 0,
            ...HelperStyle.bg_white,
        },
        shadow,
        rounded,
    });
};

export default Style;
