// react-native-minerva/src/components/skeleton/style.js
import { StyleSheet } from 'react-native';
import { Base, Spacing, Scale } from '../../index';

const style = (isIOS) =>
    StyleSheet.create({
        SkeletonCol: {
            width: '40%',
            marginTop: Spacing.spacer_1,
            marginBottom: Spacing.spacer_1,
        },
        opacityStyle: {
            opacity: 0.65,
        },
        lineStyle: {
            height: Base,
            marginTop: 0,
            marginBottom: 0,
        },
        lineSmallStyle: {
            height: Scale(10),
            marginTop: 0,
            marginBottom: 0,
        },
        heightNormalStyle: {
            height: Scale(48) + Spacing.spacer_1,
        },
        heightSmallStyle: {
            height: Spacing.spacer_6,
        },
        inputStyle: {
            borderRadius: Spacing.radius_2,
        },
        buttonStyle: {
            borderRadius: Scale(30),
        },
        switchContainer: {
            width: isIOS ? Scale(51) : Scale(34),
        },
        switch: {
            height: isIOS ? Scale(31) : Scale(27),
        },
    });

export default style;
