// react-native-minerva/src/components/emptyState/style.js
import { StyleSheet } from 'react-native';
import { Spacing, Scale } from '../../index';

const circleDefault = {
    alignSelf: 'center',
    position: 'absolute',
};

const style = StyleSheet.create({
    containerImage: {
        width: '100%',
        height: Scale(184),
        justifyContent: 'center',
        alignItems: 'center',
        marginBottom: Spacing.spacer_4,
    },
    circleSmall1: {
        ...circleDefault,
        top: 0,
        bottom: 0,
        left: Spacing.spacer_7,
        height: Spacing.spacer_6,
        width: Spacing.spacer_6,
        borderRadius: Spacing.spacer_6 / 2,
    },
    circleSmall2: {
        ...circleDefault,
        bottom: 0,
        right: Spacing.spacer_7,
        height: Spacing.spacer_6,
        width: Spacing.spacer_6,
        borderRadius: Spacing.spacer_6 / 2,
    },
    circleSmall3a: {
        ...circleDefault,
        top: 0,
        right: Spacing.spacer_7,
        height: Spacing.spacer_5,
        width: Spacing.spacer_5,
        borderRadius: Spacing.spacer_5 / 2,
    },
    circleSmall3: {
        ...circleDefault,
        top: Scale(-10),
        bottom: 0,
        right: Scale(65),
        height: Spacing.spacer_5,
        width: Spacing.spacer_5,
        borderRadius: Spacing.spacer_5 / 2,
    },
    circleMedium1: {
        ...circleDefault,
        top: Scale(10),
        right: 0,
        height: Scale(72),
        width: Scale(72),
        borderRadius: Scale(72) / 2,
    },
    circleMedium2: {
        ...circleDefault,
        top: 0,
        right: 0,
        height: Scale(72),
        width: Scale(72),
        borderRadius: Scale(72) / 2,
    },
    circleMedium3: {
        ...circleDefault,
        top: Scale(25),
        right: Scale(0),
        height: Scale(50),
        width: Scale(50),
        borderRadius: Scale(50) / 2,
    },
    circleLarge1: {
        ...circleDefault,
        left: 0,
        bottom: 0,
        right: 0,
        height: Scale(120),
        width: Scale(120),
        borderRadius: Scale(120) / 2,
    },
    circleLarge2: {
        ...circleDefault,
        left: 0,
        bottom: 0,
        height: Scale(120),
        width: Scale(120),
        borderRadius: Scale(120) / 2,
    },
    circleLarge3: {
        ...circleDefault,
        top: 0,
        left: Scale(-20),
        height: Scale(170),
        width: Scale(170),
        borderRadius: Scale(170) / 2,
    },
    circleContainer: {
        position: 'absolute',
        width: Scale(200),
        alignSelf: 'center',
        height: Scale(160),
        zIndex: -1,
        transform: [{ translateY: -Scale(50) }],
    },
});

export default style;
