// react-native-minerva/native-base/components/Badge.js
// @flow
import { Spacing, Scale, minerva } from '../../src/index';

export default (variables /* : * */ = minerva) => ({
    '.smaller': {
        height: Scale(28),
        minWidth: Scale(28),
        borderRadius: Scale(28) / 2,
    },
    '.small': {
        height: Spacing.spacer_7,
        minWidth: Spacing.spacer_7,
        borderRadius: Spacing.spacer_7 / 2,
    },
    '.large': {
        height: Scale(120),
        minWidth: Scale(120),
        borderRadius: Scale(120) / 2,
    },
    '.close': {
        backgroundColor: variables.gray,
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: Scale(2),
        borderColor: 'white',
        height: Scale(20),
        width: Scale(20),
        minWidth: Scale(20),
        borderRadius: Scale(20) / 2,
        padding: 0,
    },
    '.medium': {
        height: Scale(48),
        minWidth: Scale(48),
        borderRadius: Scale(48) / 2,
    },
    '.primary': {
        backgroundColor: variables.primary_background,
        'NativeBase.Text': {
            color: variables.primary,
        },
    },
    '.warning': {
        backgroundColor: variables.warning_background,
        'NativeBase.Text': {
            color: variables.warning,
        },
    },
    '.info': {
        backgroundColor: variables.info_background,
        'NativeBase.Text': {
            color: variables.info,
        },
    },
    '.success': {
        backgroundColor: variables.success_background,
        'NativeBase.Text': {
            color: variables.success,
        },
    },
    '.danger': {
        backgroundColor: variables.danger_background,
        'NativeBase.Text': {
            color: variables.danger,
        },
    },
    'NativeBase.Text': {
        fontSize: Spacing.type_1,
        textAlign: 'center',
        marginBottom: 0,
    },
    '.noPadding':{
        padding: 0,
    },
    backgroundColor: variables.gray_background,
    padding: variables.badgePadding,
    paddingHorizontal: Scale(8),
    justifyContent: 'center',
    borderRadius: Scale(8),
    height: Scale(27),
    minWidth: Scale(27),
});
