// react-native-minerva/native-base/components/Button.js
import { Spacing, Fonts, Scale, HelperStyle } from '../../src/index';
import variable from '../variables/minerva';

export default (variables /* : * */ = variable) => {
    const primaryCommon = {
        'NativeBase.Text': {
            color: variables.white,
        },
    };
    const secondaryCommon = {
        'NativeBase.Text': {
            color: variables.primary,
        },
    };
    const neutralCommon = {
        'NativeBase.Text': {
            color: variables.titleFontColor,
        },
    };
    const errorCommon = {
        'NativeBase.Text': {
            color: variables.danger,
        },
    };
    const inverseCommon = {
        'NativeBase.Text': {
            color: variables.white,
        },
    };

    return {
        '.disabled': {
            '.transparent': {
                backgroundColor: null,
                'NativeBase.Text': {
                    color: variables.buttonDisabledBg,
                },
            },
        },
        '.primary': {
            ...primaryCommon,
            '.disabled': {
                backgroundColor: 'transparent',
            },
        },
        '.secondary': {
            ...secondaryCommon,
            '.disabled': {
                'NativeBase.Text': {
                    color: variables.gray_light,
                },
            },
        },
        '.neutral': {
            ...neutralCommon,
            '.disabled': {
                'NativeBase.Text': {
                    color: variables.gray_light,
                },
            },
        },
        '.error': {
            ...errorCommon,
            '.disabled': {
                'NativeBase.Text': {
                    color: variables.gray_light,
                },
            },
        },
        '.inverse': {
            ...inverseCommon,
            '.disabled': {
                'NativeBase.Text': {
                    color: variables.gray_light,
                },
            },
        },
        '.small': {
            ...HelperStyle.as_fs,
            'NativeBase.Text': {
                fontSize: Spacing.type_1,
                lineHeight: Spacing.lineHeight_1,
            },
        },
        '.smaller': {
            ...HelperStyle.as_fs,
            'NativeBase.Text': {
                fontSize: Spacing.type_1,
                lineHeight: Spacing.lineHeight_1,
            },
        },
        '.circle': {
            ...HelperStyle.as_fs,
        },
        '.tiny': {
            'NativeBase.Text': {
                fontSize: Scale(10),
                lineHeight: Spacing.lineHeight_1,
                marginBottom: 0,
            },
        },
        '.full': {
            justifyContent: 'center',
            alignSelf: 'stretch',
            borderRadius: 0,
        },
        '.vertical': {
            flexDirection: 'column',
            height: null,
        },
        'NativeBase.Text': {
            fontFamily: Fonts.fontFamilySecondaryBold,
            textTransform: 'uppercase',
            fontSize: Spacing.type_2,
            lineHeight: Spacing.lineHeight_2,
            letterSpacing: 0.4,
            marginLeft: 0,
            marginRight: 0,
            marginBottom: 0,
            color: variables.inverseTextColor,
            backgroundColor: 'transparent',
        },
        '.link': {
            backgroundColor: 'transparent',
            padding: 0,
            margin: 0,
            justifyContent: 'flex-start',
            ...HelperStyle.as_fs,
            '.small': {
                'NativeBase.Text': {
                    lineHeight: Spacing.lineHeight_2,
                },
            },
            '.smaller': {
                'NativeBase.Text': {
                    lineHeight: Spacing.lineHeight_2,
                },
            },
            'NativeBase.Text': {
                textDecorationLine: 'underline',
            },
        },
        '.footerTab': {
            flex: 1,
            flexDirection: 'column',
            height: Scale(64),
            backgroundColor: 'transparent',
            justifyContent: 'center',
            alignContent: 'center',
            'NativeBase.Text': {
                color: variables.gray,
                textTransform: 'capitalize',
                fontFamily: variables.fontFamily,
            },
        },
        '.active': {
            'NativeBase.Text': {
                color: variables.primary,
            },
        },
        paddingTop: 0,
        paddingBottom: 0,
        paddingLeft: 0,
        paddingRight: 0,
        backgroundColor: 'transparent',
        borderRadius: variables.borderRadiusLarge,
        borderColor: variables.buttonPrimaryBg,
        borderWidth: null,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
    };
};
