// react-native-minerva/native-base/components/CardItem.js
// @flow

import { StyleSheet } from 'react-native';
import { Spacing, Scale } from '../../src/index';
import variable from '../variables/platform';
import { PLATFORM } from '../variables/commonColor';

export default (variables /* : * */ = variable) => {
    const { platform } = variables;

    const TypeMarginBottomNone = {
        'NativeBase.Text': {
            marginBottom: 0,
        },
        'NativeBase.H1': {
            marginBottom: 0,
        },
        'NativeBase.H2': {
            marginBottom: 0,
        },
        'NativeBase.H3': {
            marginBottom: 0,
        },
    };

    const transparentBtnCommon = {
        'NativeBase.Text': {
            fontSize: variables.DefaultFontSize - 3,
            color: variables.sTabBarActiveTextColor,
        },
        'NativeBase.Icon': {
            fontSize: variables.iconFontSize - Scale(10),
            color: variables.sTabBarActiveTextColor,
            marginHorizontal: null,
        },
        'NativeBase.IconNB': {
            fontSize: variables.iconFontSize - Scale(10),
            color: variables.sTabBarActiveTextColor,
        },
        paddingVertical: null,
        paddingHorizontal: null,
    };

    return {
        'NativeBase.Left': {
            'NativeBase.Body': {
                flex: 1,
                marginLeft: Scale(10),
                alignItems: null,
            },
            'NativeBase.Icon': {
                fontSize: variables.iconFontSize,
            },
            'NativeBase.IconNB': {
                fontSize: variables.iconFontSize,
            },
            'NativeBase.Text': {
                marginLeft: Scale(10),
                alignSelf: 'center',
            },
            'NativeBase.Button': {
                '.transparent': {
                    ...transparentBtnCommon,
                    paddingRight: variables.cardItemPadding + Scale(5),
                },
            },
            flex: 1,
            flexDirection: 'row',
            alignItems: 'center',
        },
        '.content': {
            'NativeBase.Text': {
                color: platform === PLATFORM.IOS ? '#555' : '#222',
                fontSize: variables.DefaultFontSize - 2,
            },
        },
        '.body': {
            '.small': {
                paddingTop: Spacing.spacer_0,
                paddingLeft: Spacing.spacer_0,
                paddingRight: Spacing.spacer_0,
                paddingBottom: Spacing.spacer_3,
            },
            flexDirection: null,
            alignItems: null,
            paddingTop: Spacing.spacer_1,
            paddingBottom: Spacing.spacer_4,
            paddingHorizontal: Spacing.spacer_4,
        },
        'NativeBase.Body': {
            'NativeBase.Button': {
                '.transparent': {
                    ...transparentBtnCommon,
                    paddingRight: variables.cardItemPadding + Scale(5),
                    alignSelf: 'stretch',
                },
            },
            flex: 1,
            alignSelf: 'stretch',
            alignItems: 'flex-start',
        },
        'NativeBase.Right': {
            'NativeBase.Badge': {
                alignSelf: null,
            },
            'NativeBase.Button': {
                '.transparent': {
                    ...transparentBtnCommon,
                },
                alignSelf: null,
            },
            'NativeBase.Icon': {
                alignSelf: null,
                fontSize: variables.iconFontSize - Spacing.spacer_2,
                color: variables.cardBorderColor,
            },
            'NativeBase.IconNB': {
                alignSelf: null,
                fontSize: variables.iconFontSize - Spacing.spacer_2,
                color: variables.cardBorderColor,
            },
            'NativeBase.Text': {
                fontSize: variables.DefaultFontSize - 1,
                alignSelf: null,
            },
            'NativeBase.Thumbnail': {
                alignSelf: null,
            },
            'NativeBase.Image': {
                alignSelf: null,
            },
            'NativeBase.Radio': {
                alignSelf: null,
            },
            'NativeBase.Checkbox': {
                alignSelf: null,
            },
            'NativeBase.Switch': {
                alignSelf: null,
            },
            flex: 0.8,
        },
        '.header': {
            'NativeBase.Body': {
                ...TypeMarginBottomNone,
            },
            'NativeBase.Left': {
                ...TypeMarginBottomNone,
            },
            'NativeBase.Right': {
                ...TypeMarginBottomNone,
            },
            ...TypeMarginBottomNone,
            '.bordered': {
                'NativeBase.Text': {
                    color: variables.brandPrimary,
                },
                borderBottomWidth: variables.borderWidth,
            },
            '.small': {
                paddingTop: Spacing.spacer_4,
            },
            borderWidth: 0,
            borderColor: 'transparent',
            borderTopLeftRadius: Spacing.radius_3,
            borderTopRightRadius: Spacing.radius_3,
            paddingTop: Spacing.spacer_5,
            paddingBottom: Spacing.spacer_4,
            paddingHorizontal: Spacing.spacer_4,
        },
        '.footer': {
            'NativeBase.Text': {
                fontSize: Spacing.type_3,
            },
            '.bordered': {
                'NativeBase.Text': {
                    color: variables.brandPrimary,
                },
                borderTopWidth: variables.borderWidth,
            },
            borderWidth: 0,
            borderColor: 'transparent',
            borderBottomLeftRadius: Spacing.radius_3,
            borderBottomRightRadius: Spacing.radius_3,
            paddingVertical: Spacing.spacer_4,
            paddingHorizontal: Spacing.spacer_4,
            flexDirection: null,
            alignItems: null,
        },
        'NativeBase.Icon': {
            width: variables.iconFontSize + Scale(5),
            fontSize: variables.iconFontSize - 2,
        },
        'NativeBase.IconNB': {
            width: variables.iconFontSize + Scale(5),
            fontSize: variables.iconFontSize - 2,
        },
        '.bordered': {
            borderBottomWidth: StyleSheet.hairlineWidth,
            borderColor: variables.cardBorderColor,
        },
        '.first': {
            borderTopLeftRadius: variables.cardBorderRadius,
            borderTopRightRadius: variables.cardBorderRadius,
        },
        '.last': {
            borderBottomLeftRadius: variables.cardBorderRadius,
            borderBottomRightRadius: variables.cardBorderRadius,
        },
        '.avatar': {
            'NativeBase.Left': {
                flex: 0,
                alignSelf: 'center',
                marginRight: variables.cardItemPadding * 2,
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
                flex: 1,
                justifyContent: 'center',
            },
            'NativeBase.Right': {
                'NativeBase.Text': {
                    marginBottom: 0,
                    '.note': {
                        fontSize: variables.noteFontSize - 2,
                    },
                },
                flex: 0,
                alignSelf: 'stretch',
                justifyContent: 'center',
            },
            '.noBorder': {
                'NativeBase.Body': {
                    borderBottomWidth: null,
                },
                'NativeBase.Right': {
                    borderBottomWidth: null,
                },
            },
            paddingVertical: Spacing.spacer_4,
        },
        '.row': {
            flexDirection: null,
            alignItems: null,
        },
        flexDirection: 'row',
        alignItems: 'center',
        paddingVertical: Spacing.spacer_4,
        paddingHorizontal: Spacing.spacer_4,
    };
};
