// react-native-minerva/native-base/components/Card.js
// @flow

import { Spacing, HelperStyle } from '../../src/index';
import variable from '../variables/minerva';

export default (variables /* : * */ = variable) => ({
    '.transparent': {
        shadowColor: null,
        shadowOffset: null,
        shadowOpacity: null,
        shadowRadius: null,
        elevation: null,
        backgroundColor: 'transparent',
        borderWidth: 0,
    },
    '.noShadow': {
        shadowColor: null,
        shadowOffset: null,
        shadowOpacity: null,
        elevation: null,
    },
    '.alert': {
        padding: 0,
        margin: 0,
        backgroundColor: 'transparent',
        borderColor: 'transparent',
        'NativeBase.CardItem': {
            borderWidth: 1,
            borderLeftWidth: 16,
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
            },
            'NativeBase.Right': {
                'NativeBase.Text': {
                    color: 'white',
                },
                'NativeBase.Body': {
                    'NativeBase.Text': {
                        color: 'white',
                    },
                },
            },
            'NativeBase.Left': {
                'NativeBase.Text': {
                    color: 'white',
                },
                'NativeBase.Body': {
                    'NativeBase.Text': {
                        color: 'white',
                    },
                },
            },
            backgroundColor: 'transparent',
            padding: Spacing.spacer_4,
            borderTopLeftRadius: 0,
            borderBottomLeftRadius: 0,
            borderTopRightRadius: Spacing.radius_2,
            borderBottomRightRadius: Spacing.radius_2,
        },
        marginBottom: Spacing.spacer_4,
        marginRight: Spacing.spacer_4,
    },
    '.success': {
        'NativeBase.CardItem': {
            backgroundColor: variables.success_background,
            borderColor: variables.success_light,
            'NativeBase.Left': {
                'NativeBase.Icon': {
                    color: variables.success,
                },
                'NativeBase.LottieIcon': {
                    'NativeBase.Text': {
                        color: variables.success,
                    },
                },
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    color: variables.success,
                },
            },
        },
    },
    '.info': {
        'NativeBase.CardItem': {
            backgroundColor: variables.info_background,
            borderColor: variables.info_light,
            'NativeBase.Left': {
                'NativeBase.Icon': {
                    color: variables.info,
                },
                'NativeBase.LottieIcon': {
                    'NativeBase.Text': {
                        color: variables.info,
                    },
                },
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    color: variables.info,
                },
            },
        },
    },
    '.warning': {
        'NativeBase.CardItem': {
            backgroundColor: variables.warning_background,
            borderColor: variables.warning_light,
            'NativeBase.Left': {
                'NativeBase.Icon': {
                    color: variables.warning,
                },
                'NativeBase.LottieIcon': {
                    'NativeBase.Text': {
                        color: variables.warning,
                    },
                },
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    color: variables.warning,
                },
            },
        },
    },
    '.danger': {
        'NativeBase.CardItem': {
            backgroundColor: variables.danger_background,
            borderColor: variables.danger_light,
            'NativeBase.Left': {
                'NativeBase.Icon': {
                    color: variables.danger,
                },
                'NativeBase.LottieIcon': {
                    'NativeBase.Text': {
                        color: variables.danger,
                    },
                },
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    color: variables.danger,
                },
            },
        },
    },
    '.neutral': {
        'NativeBase.CardItem': {
            backgroundColor: variables.gray_background,
            borderColor: variables.gray_light,
            'NativeBase.Left': {
                'NativeBase.Icon': {
                    color: variables.gray,
                },
                'NativeBase.LottieIcon': {
                    'NativeBase.Text': {
                        color: variables.gray,
                    },
                },
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    color: variables.gray,
                },
            },
        },
    },
    '.square': {
        borderRadius: 0,
    },
    '.bordered': {
        borderWidth: 0.5,
        borderColor: variables.cardBorderColor,
    },
    '.firstItem': {
        paddingTop: Spacing.spacer_2,
        borderBottomLeftRadius: Spacing.border_0,
        borderBottomWidth: 0,
        borderBottomRightRadius: Spacing.border_0,
        marginBottom: 0,
        '.lineSeparator': {
            borderBottomWidth: 0.5,
        }
    },
    '.middleItem': {
        borderTopLeftRadius: Spacing.border_0,
        borderTopRightRadius: Spacing.border_0,
        borderBottomRightRadius: Spacing.border_0,
        borderBottomLeftRadius: Spacing.border_0,
        borderTopWidth: 0,
        borderBottomWidth: 0,
        marginBottom: 0,
        marginTop: 0,
        '.lineSeparator': {
            borderTopWidth: 0.5,
            borderBottomWidth: 0.5,
        }
    },
    '.lastItem': {
        marginBottom: Spacing.spacer_2,
        paddingBottom: Spacing.spacer_2,
        borderTopLeftRadius: Spacing.border_0,
        borderTopWidth: 0,
        borderTopRightRadius: Spacing.border_0,
        '.lineSeparator': {
            borderTopWidth: 0.5,
        }
    },
    borderWidth: 0,
    borderRadius: variables.cardBorderRadius,
    borderColor: 'transparent',
    flexWrap: 'nowrap',
    marginBottom: Spacing.spacer_4,
    marginHorizontal: 0,
    backgroundColor: variables.cardDefaultBg,
    ...HelperStyle.shadow,
});
