// react-native-minerva/native-base/components/CheckBox.js
// @flow
import minerva from '../variables/minerva';

export default (variables /* : * */ = minerva) => ({
    '.checked': {
        'NativeBase.Icon': {
            color: variables.checkboxTickColor,
        },
        'NativeBase.IconNB': {
            color: variables.checkboxTickColor,
        },
        borderColor: variables.CheckboxBorderColorchecked,
        '.disabled': {
            backgroundColor: variables.primary_lighter,
        },
    },
    'NativeBase.Icon': {
        color: 'transparent',
        lineHeight: variables.CheckboxIconSize,
        marginTop: variables.CheckboxIconMarginTop,
        fontSize: variables.CheckboxFontSize,
    },
    'NativeBase.IconNB': {
        color: 'transparent',
        lineHeight: variables.CheckboxIconSize,
        marginTop: variables.CheckboxIconMarginTop,
        fontSize: variables.CheckboxFontSize,
    },
    '.disabled': {
        borderColor: variables.CheckboxBorderColorDisabled,
    },
    borderRadius: variables.CheckboxRadius,
    overflow: 'hidden',
    width: variables.checkboxSize,
    height: variables.checkboxSize,
    borderWidth: variables.CheckboxBorderWidth,
    borderColor: variables.CheckboxBorderColor,
    paddingLeft: variables.CheckboxPaddingLeft - 1,
    paddingBottom: variables.CheckboxPaddingBottom,
    left: 0,
    alignItems: 'center',
});
