// react-native-minerva/native-base/components/Footer.js
// @flow

import { minerva, Spacing } from '../../src/index';

export default (variables /* : * */ = minerva) => {
    const iconCommon = {
        'NativeBase.Icon': {
            color: variables.tabBarActiveTextColor,
        },
    };
    const iconNBCommon = {
        'NativeBase.IconNB': {
            color: variables.tabBarActiveTextColor,
        },
    };
    const textCommon = {
        'NativeBase.Text': {
            color: variables.tabBarActiveTextColor,
        },
    };
    return {
        'NativeBase.Left': {
            'NativeBase.Button': {
                '.transparent': {
                    backgroundColor: 'transparent',
                    borderColor: null,
                    elevation: 0,
                    shadowColor: null,
                    shadowOffset: null,
                    shadowRadius: null,
                    shadowOpacity: null,
                    ...iconCommon,
                    ...iconNBCommon,
                    ...textCommon,
                },
                alignSelf: null,
                ...iconCommon,
                ...iconNBCommon,
                // ...textCommon
            },
            flex: 1,
            alignSelf: 'center',
            alignItems: 'flex-start',
        },
        'NativeBase.Body': {
            flex: 1,
            alignItems: 'center',
            alignSelf: 'center',
            flexDirection: 'row',
            'NativeBase.Button': {
                alignSelf: 'center',
                '.transparent': {
                    backgroundColor: 'transparent',
                    borderColor: null,
                    elevation: 0,
                    shadowColor: null,
                    shadowOffset: null,
                    shadowRadius: null,
                    shadowOpacity: null,
                    ...iconCommon,
                    ...iconNBCommon,
                    ...textCommon,
                },
                '.full': {
                    height: variables.footerHeight,
                    paddingBottom: variables.footerPaddingBottom,
                    flex: 1,
                },
                ...iconCommon,
                ...iconNBCommon,
                // ...textCommon
            },
        },
        'NativeBase.Right': {
            'NativeBase.Button': {
                '.transparent': {
                    backgroundColor: 'transparent',
                    borderColor: null,
                    elevation: 0,
                    shadowColor: null,
                    shadowOffset: null,
                    shadowRadius: null,
                    shadowOpacity: null,
                    ...iconCommon,
                    ...iconNBCommon,
                    ...textCommon,
                },
                alignSelf: null,
                ...iconCommon,
                ...iconNBCommon,
                // ...textCommon
            },
            flex: 1,
            alignSelf: 'center',
            alignItems: 'flex-end',
        },
        '.footerMenuHome': {
            paddingHorizontal: Spacing.spacer_4,
            borderTopLeftRadius: Spacing.radius_3,
            borderTopRightRadius: Spacing.radius_3,
            borderBottomLeftRadius: 0,
            borderBottomRightRadius: 0,
            backgroundColor: variables.white,
            borderWidth: 0,
            borderColor: 'transparent',
            shadowColor: '#000',
            shadowOffset: { width: 0, height: -3 },
            shadowOpacity: 0.05,
            shadowRadius: 2,
            elevation: 10,
        },
        backgroundColor: variables.footerDefaultBg,
        flexDirection: 'row',
        justifyContent: 'center',
        paddingBottom: variables.footerPaddingBottom,
        elevation: 0,
        left: 0,
        right: 0,
    };
};
