// react-native-minerva/native-base/components/Form.js
// @flow
import { Scale } from '../../src/index';

export default () => ({
    'NativeBase.Item': {
        '.fixedLabel': {
            'NativeBase.Label': {
                paddingLeft: null,
            },
            marginLeft: Scale(15),
        },
        '.inlineLabel': {
            'NativeBase.Label': {
                paddingLeft: null,
            },
            marginLeft: Scale(15),
        },
        '.placeholderLabel': {
            'NativeBase.Input': {},
        },
        '.stackedLabel': {
            'NativeBase.Label': {
                top: Scale(5),
                paddingLeft: null,
            },
            'NativeBase.Input': {
                paddingLeft: null,
                marginLeft: null,
            },
            'NativeBase.Icon': {
                marginTop: Scale(36),
            },
            marginLeft: Scale(15),
        },
        '.floatingLabel': {
            'NativeBase.Input': {
                paddingLeft: null,
                top: Scale(10),
                marginLeft: null,
            },
            'NativeBase.Label': {
                left: 0,
                top: Scale(6),
            },
            'NativeBase.Icon': {
                top: Scale(6),
            },
            marginTop: Scale(15),
            marginLeft: Scale(15),
        },
        '.regular': {
            'NativeBase.Label': {
                left: 0,
            },
            marginLeft: 0,
        },
        '.rounded': {
            'NativeBase.Label': {
                left: 0,
            },
            marginLeft: 0,
        },
        '.underline': {
            'NativeBase.Label': {
                left: 0,
                top: 0,
                position: 'relative',
            },
            'NativeBase.Input': {
                left: -Scale(15),
            },
            marginLeft: Scale(15),
        },
        '.last': {
            marginLeft: 0,
            paddingLeft: Scale(15),
        },
        'NativeBase.Label': {
            paddingRight: Scale(5),
        },
        marginLeft: Scale(15),
    },
});
