// react-native-minerva/native-base/components/Header.js
/* eslint-disable sonarjs/cognitive-complexity */
// @flow

import { PixelRatio, StatusBar } from 'react-native';
import { Spacing, Scale, HelperStyle } from '../../src/index';
import variable from '../variables/platform';
import { PLATFORM } from '../variables/commonColor';

export default (variables /* : * */ = variable) => {
    const { platformStyle } = variables;
    const { platform } = variables;
    return {
        '.span': {
            height: 128,
            'NativeBase.Left': {
                ...HelperStyle.as_fs,
            },
            'NativeBase.Body': {
                alignSelf: 'flex-end',
                ...HelperStyle.ai_fs,
                justifyContent: 'center',
                paddingBottom: Scale(26),
            },
            'NativeBase.Right': {
                ...HelperStyle.as_fs,
            },
        },
        '.hasSubtitle': {
            'NativeBase.Body': {
                'NativeBase.Title': {
                    fontSize: variables.titleFontSize - 2,
                    fontFamily: variables.titleFontfamily,
                    textAlign: 'center',
                    fontWeight: '500',
                    paddingBottom: Scale(3),
                },
                'NativeBase.Subtitle': {
                    fontSize: variables.subTitleFontSize,
                    fontFamily: variables.titleFontfamily,
                    color: variables.subtitleColor,
                    textAlign: 'center',
                },
            },
        },
        '.transparent': {
            backgroundColor: 'transparent',
            borderBottomColor: 'transparent',
            elevation: 0,
            shadowColor: null,
            shadowOffset: null,
            shadowRadius: null,
            shadowOpacity: null,
            paddingTop: platform === PLATFORM.ANDROID ? StatusBar.currentHeight : undefined,
            height:
                platform === PLATFORM.ANDROID
                    ? variables.toolbarHeight + StatusBar.currentHeight
                    : variables.toolbarHeight,
        },
        '.noShadow': {
            elevation: 0,
            shadowColor: null,
            shadowOffset: null,
            shadowRadius: null,
            shadowOpacity: null,
        },
        '.hasTabs': {
            elevation: 0,
            shadowColor: null,
            shadowOffset: null,
            shadowRadius: null,
            shadowOpacity: null,
            borderBottomWidth: null,
        },
        '.hasSegment': {
            elevation: 0,
            shadowColor: null,
            shadowOffset: null,
            shadowRadius: null,
            shadowOpacity: null,
            borderBottomWidth: null,
            'NativeBase.Left': {
                flex: 0.3,
            },
            'NativeBase.Right': {
                flex: 0.3,
            },
            'NativeBase.Body': {
                flex: 1,
                'NativeBase.Segment': {
                    marginRight: 0,
                    alignSelf: 'center',
                    'NativeBase.Button': {
                        paddingLeft: 0,
                        paddingRight: 0,
                    },
                },
            },
        },
        '.noLeft': {
            'NativeBase.Left': {
                width: platform === PLATFORM.IOS ? undefined : 0,
                flex: platform === PLATFORM.IOS ? 1 : 0,
            },
            'NativeBase.Body': {
                'NativeBase.Title': {
                    paddingLeft: platform === PLATFORM.IOS ? undefined : Scale(10),
                },
                'NativeBase.Subtitle': {
                    paddingLeft: platform === PLATFORM.IOS ? undefined : Scale(10),
                },
            },
        },
        'NativeBase.Button': {
            justifyContent: 'center',
            alignSelf: 'center',
            alignItems: 'center',
            '.transparent': {
                'NativeBase.Text': {
                    color: variables.toolbarBtnTextColor,
                    fontWeight: '600',
                },
                'NativeBase.Icon': {
                    color: variables.toolbarBtnColor,
                },
                'NativeBase.IconNB': {
                    color: variables.toolbarBtnColor,
                },
                paddingHorizontal: variables.buttonPadding,
            },
            paddingHorizontal: Scale(15),
        },
        '.searchBar': {
            'NativeBase.Item': {
                'NativeBase.Icon': {
                    backgroundColor: 'transparent',
                    color: variables.dropdownLinkColor,
                    fontSize: variables.toolbarSearchIconSize,
                    alignItems: 'center',
                    marginTop: Scale(2),
                    paddingRight: Scale(10),
                    paddingLeft: Scale(10),
                },
                'NativeBase.IconNB': {
                    backgroundColor: 'transparent',
                    color: null,
                    alignSelf: 'center',
                },
                'NativeBase.Input': {
                    alignSelf: 'center',
                    lineHeight: null,
                    height: variables.searchBarInputHeight,
                },
                alignSelf: 'center',
                alignItems: 'center',
                justifyContent: 'flex-start',
                flex: 1,
                height: variables.searchBarHeight,
                borderColor: 'transparent',
                backgroundColor: variables.toolbarInputColor,
            },
            'NativeBase.Button': {
                '.transparent': {
                    'NativeBase.Text': {
                        fontWeight: '500',
                    },
                    paddingHorizontal: null,
                    paddingLeft: platform === PLATFORM.IOS ? Scale(10) : null,
                },
                paddingHorizontal: platform === PLATFORM.IOS ? undefined : null,
                width: platform === PLATFORM.IOS ? undefined : 0,
                height: platform === PLATFORM.IOS ? undefined : 0,
            },
        },
        '.rounded': {
            'NativeBase.Item': {
                borderRadius: platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? Scale(25) : Scale(3),
            },
        },
        'NativeBase.Left': {
            'NativeBase.Button': {
                '.hasText': {
                    marginLeft: -Scale(10),
                    height: Scale(30),
                    'NativeBase.Icon': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize,
                        marginTop: Scale(2),
                        marginRight: Scale(5),
                        marginLeft: Scale(2),
                    },
                    'NativeBase.Text': {
                        color: variables.toolbarBtnTextColor,
                        fontSize: platform === PLATFORM.IOS ? 17 : 0,
                        marginLeft: Scale(7),
                        lineHeight: Scale(19, 5),
                    },
                    'NativeBase.IconNB': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize,
                        marginTop: Scale(2),
                        marginRight: Scale(5),
                        marginLeft: Scale(2),
                    },
                },
                '.transparent': {
                    marginLeft: platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? -Scale(3) : 0,
                    'NativeBase.Icon': {
                        color: variables.toolbarBtnColor,
                        fontSize:
                            platform === PLATFORM.IOS && variables.platformStyle !== PLATFORM.MATERIAL
                                ? variables.iconHeaderSize + 1
                                : variables.iconHeaderSize,
                        marginTop: 0,
                        marginRight: Scale(2),
                        marginLeft: 1,
                        paddingTop: 1,
                    },
                    'NativeBase.IconNB': {
                        color: variables.toolbarBtnColor,
                        fontSize:
                            platform === PLATFORM.IOS && variables.platformStyle !== PLATFORM.MATERIAL
                                ? variables.iconHeaderSize + 1
                                : variables.iconHeaderSize - Scale(2),
                        marginTop: 0,
                        marginRight: Scale(2),
                        marginLeft: 1,
                        paddingTop: 1,
                    },
                    'NativeBase.Text': {
                        color: variables.toolbarBtnTextColor,
                        fontSize: platform === PLATFORM.IOS ? Scale(17) : 0,
                        top: platform === PLATFORM.IOS ? 1 : -1.5,
                        paddingLeft:
                            platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? Scale(2) : Scale(5),
                        paddingRight:
                            platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? undefined : Scale(10),
                    },
                    backgroundColor: 'transparent',
                    borderColor: null,
                    elevation: 0,
                    shadowColor: null,
                    shadowOffset: null,
                    shadowRadius: null,
                    shadowOpacity: null,
                },
                'NativeBase.Icon': {
                    color: variables.toolbarBtnColor,
                },
                'NativeBase.IconNB': {
                    color: variables.toolbarBtnColor,
                },
                alignSelf: null,
                paddingRight: variables.buttonPadding,
                paddingLeft:
                    platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL
                        ? Spacing.spacer_1
                        : Spacing.spacer_2,
            },
            flex: platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? 1 : 0.4,
            alignSelf: 'center',
            ...HelperStyle.ai_fs,
        },
        'NativeBase.Body': {
            flex: 1,
            alignItems: platform === PLATFORM.IOS && platformStyle !== PLATFORM.MATERIAL ? 'center' : 'flex-start',
            alignSelf: 'center',
            'NativeBase.Segment': {
                borderWidth: 0,
                alignSelf: 'flex-end',
                marginRight: platform === PLATFORM.IOS ? -Spacing.spacer_7 : -Scale(55),
            },
            'NativeBase.Button': {
                alignSelf: 'center',
                '.transparent': {
                    backgroundColor: 'transparent',
                },
                'NativeBase.Icon': {
                    color: variables.toolbarBtnColor,
                },
                'NativeBase.IconNB': {
                    color: variables.toolbarBtnColor,
                },
                'NativeBase.Text': {
                    color: variables.inverseTextColor,
                    backgroundColor: 'transparent',
                },
            },
        },
        'NativeBase.Right': {
            'NativeBase.Button': {
                '.hasText': {
                    height: Scale(30),
                    'NativeBase.Icon': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize - Scale(2),
                        marginTop: Scale(2),
                        marginRight: Scale(2),
                        marginLeft: Scale(5),
                    },
                    'NativeBase.Text': {
                        color: variables.toolbarBtnTextColor,
                        fontSize: platform === PLATFORM.IOS ? Scale(17) : Spacing.type_2,
                        lineHeight: Scale(19.5),
                    },
                    'NativeBase.IconNB': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize - Scale(2),
                        marginTop: Scale(2),
                        marginRight: Scale(2),
                        marginLeft: Scale(5),
                    },
                },
                '.transparent': {
                    marginRight: platform === PLATFORM.IOS ? -Scale(9) : -Scale(5),
                    paddingLeft: Scale(15),
                    paddingRight: Spacing.spacer_3,
                    paddingHorizontal: Scale(15),
                    borderRadius: Scale(50),
                    'NativeBase.Icon': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize - Scale(2),
                        marginTop: 0,
                        marginLeft: Scale(2),
                        marginRight: 0,
                        // paddingTop: 0
                    },
                    'NativeBase.IconNB': {
                        color: variables.toolbarBtnColor,
                        fontSize: variables.iconHeaderSize - Scale(2),
                        marginTop: 0,
                        marginLeft: Scale(2),
                        marginRight: 0,
                        // paddingTop: 0
                    },
                    'NativeBase.Text': {
                        color: variables.toolbarBtnTextColor,
                        fontSize: platform === PLATFORM.IOS ? Scale(17) : Spacing.type_2,
                        top: platform === PLATFORM.IOS ? 1 : -1.5,
                        paddingRight:
                            platform === PLATFORM.IOS && variables.platformStyle !== PLATFORM.MATERIAL ? 0 : undefined,
                    },
                    backgroundColor: 'transparent',
                    borderColor: null,
                    elevation: 0,
                    shadowColor: null,
                    shadowOffset: null,
                    shadowRadius: null,
                    shadowOpacity: null,
                },
                'NativeBase.Icon': {
                    color: variables.toolbarBtnColor,
                },
                'NativeBase.IconNB': {
                    color: variables.toolbarBtnColor,
                },
                alignSelf: null,
                paddingHorizontal: variables.buttonPadding,
            },
            flex: 1,
            alignSelf: 'center',
            alignItems: 'flex-end',
            flexDirection: 'row',
            justifyContent: 'flex-end',
        },
        backgroundColor: variables.toolbarDefaultBg,
        flexDirection: 'row',
        // paddingHorizontal: Scale(10),
        paddingLeft: platform === PLATFORM.IOS && variables.platformStyle !== PLATFORM.MATERIAL ? Scale(6) : Scale(10),
        paddingRight: Scale(10),
        justifyContent: 'center',
        paddingTop: platform === PLATFORM.IOS ? Scale(18) : 0,
        borderBottomWidth: platform === PLATFORM.IOS ? 1 / PixelRatio.getPixelSizeForLayoutSize(1) : 0,
        borderBottomColor: variables.toolbarDefaultBorder,
        height:
            variables.platform === PLATFORM.IOS && variables.platformStyle === PLATFORM.MATERIAL
                ? variables.toolbarHeight + Scale(10)
                : variables.toolbarHeight,
        elevation: 3,
        shadowColor: platformStyle === PLATFORM.MATERIAL ? '#000' : undefined,
        shadowOffset: platformStyle === PLATFORM.MATERIAL ? { width: 0, height: 2 } : undefined,
        shadowOpacity: platformStyle === PLATFORM.MATERIAL ? 0.2 : undefined,
        shadowRadius: platformStyle === PLATFORM.MATERIAL ? 1.2 : undefined,
        top: 0,
        left: 0,
        right: 0,
    };
};
