// react-native-minerva/native-base/components/Item.js
// @flow

import {
    Platform,
    Spacing,
    minerva,
    Base,
    Scale,
} from '../../src/index';

export default (variables /* : * */ = minerva) => {
    return {
        '.regular': {
            '.success': {
                borderColor: 'transparent',
            },
            '.error': {
                borderColor: 'transparent',
                borderWidth: 1,
            },
            borderWidth: 1,
            borderColor: 'transparent',
            padding: Base,
        },
        'NativeBase.Label': {
            fontSize: variables.inputFontSize,
            color: variables.inputColorPlaceholder,
            paddingRight: Scale(5),
        },
        'NativeBase.Icon': {
            fontSize: Spacing.type_5,
            paddingRight: 0,
        },
        'NativeBase.IconNB': {
            fontSize: Spacing.type_5,
            paddingRight: 0,
        },
        'NativeBase.Input': {
            '.multiline': {
                height: null,
            },
            height: variables.inputHeightBase,
            color: variables.inputColor,
            flex: 1,
            top: Platform.ios ? 1.5 : undefined,
            fontSize: variables.inputFontSize,
            fontFamily: variables.fontFamily,
        },
        '.success': {
            'NativeBase.Icon': {
                color: variables.inputSuccessBorderColor,
            },
            'NativeBase.IconNB': {
                color: variables.inputSuccessBorderColor,
            },
            '.rounded': {
                borderRadius: Scale(30),
                borderColor: 'transparent',
            },
            '.regular': {
                borderColor: 'transparent',
            },
            '.underline': {
                borderWidth: 1,
                borderTopWidth: 0,
                borderRightWidth: 0,
                borderLeftWidth: 0,
                borderColor: 'transparent',
            },
            borderColor: 'transparent',
        },
        '.error': {
            'NativeBase.Icon': {
                color: variables.inputErrorBorderColor,
            },
            'NativeBase.IconNB': {
                color: variables.inputErrorBorderColor,
            },
            '.rounded': {
                borderRadius: Scale(30),
                borderColor: 'transparent',
            },
            '.regular': {
                borderColor: 'transparent',
            },
            '.underline': {
                borderWidth: 1,
                borderTopWidth: 0,
                borderRightWidth: 0,
                borderLeftWidth: 0,
                borderColor: 'transparent',
            },
            borderColor: 'transparent',
        },
        '.disabled': {
            'NativeBase.Icon': {
                color: '#384850',
            },
            'NativeBase.IconNB': {
                color: '#384850',
            },
        },
        '.picker': {},
        '.inputFloatingLabel': {
            borderRadius: Spacing.radius_2,
            width: '100%',
            'NativeBase.Label': {
                lineHeight: Spacing.lineHeight_2,
                alignSelf: 'flex-start',
                paddingLeft: 0,
                paddingRight: 0,
                fontSize: Spacing.type_2,
                color: 'red',
            },
            'NativeBase.H3': {
                // fix alineacion texto
                marginBottom: 0,
                lineHeight: Scale(25),
            },
            'NativeBase.Text': {
                // fix alineacion texto
                marginBottom: 0,
                lineHeight: Scale(22),
            },
        },
        '.number': {
            'NativeBase.H3': {
                color: variables.gray_light,
            },
            'NativeBase.Input': {
                fontSize: Spacing.type_4,
                fontFamily: variables.fontFamilyBold,
                fontWeight: Platform.ios ? 'bold' : 'normal',
                letterSpacing: 1,
            },
        },
        '.pass': {
            'NativeBase.Input': {
                fontSize: Spacing.type_4,
                fontFamily: variables.fontFamilyBold,
                fontWeight: Platform.ios ? 'bold' : 'normal',
                letterSpacing: 0,
            },
        },
        '.large': {
            'NativeBase.Input': {
                fontSize: variables.fontSizeH3,
                lineHeight: variables.inputLineHeight,
            },
        },
        '.isFocus': {
            borderColor: 'transparent',
            borderWidth: 1,
            '.error': {
                borderColor: 'transparent',
            },
        },
        '.search': {
            backgroundColor: variables.gray_background,
            borderColor: 'transparent',
            marginRight: 0,
            marginLeft: 0,
            borderTopWidth: 0,
            borderRightWidth: 0,
            borderLeftWidth: 0,
            borderBottomWidth: 0,
        },
        '.small': {
            height: Spacing.spacer_7,
            borderRadius: Spacing.spacer_7 / 4,
            'NativeBase.Input': {
                height: Spacing.spacer_7,
                fontSize: Spacing.type_2,
                paddingRight: Spacing.spacer_4,
                paddingLeft: Spacing.spacer_4,
            },
        },
        '.iconRight': {
            'NativeBase.Input': {
                paddingRight: Spacing.spacer_3,
            },
        },
        '.iconLeft': {
            'NativeBase.Input': {
                paddingLeft: Spacing.spacer_3,
            },
        },
        '.inverse': {
            'NativeBase.Label': {
                color: variables.white,
            },
            'NativeBase.Input': {
                color: variables.white,
            },
        },
        fontFamily: variables.fontFamily,
        borderWidth: 1,
        borderTopWidth: 0,
        borderRightWidth: 0,
        borderLeftWidth: 0,
        borderColor: 'transparent',
        backgroundColor: 'transparent',
        flexDirection: 'row',
        alignItems: 'center',
        marginLeft: Scale(2),
        height: Scale(56),
    };
};
