// react-native-minerva/native-base/components/ListItem.js
// @flow

import { Platform, PixelRatio } from 'react-native';

import {
    Spacing,
    minerva,
    getStatusBarHeight,
    getHeaderHeight,
    Base,
    Scale,
} from '../../src/index';
import pickerTheme from './Picker';
import { PLATFORM } from '../variables/commonColor';

export default (variables /* : * */ = minerva) => {
    const { platform } = variables;
    const selectedStyle = {
        'NativeBase.Text': {
            color: variables.listItemSelected,
        },
        'NativeBase.Icon': {
            color: variables.listItemSelected,
        },
    };

    const textWhite = {
        'NativeBase.Text': {
            color: 'white',
        },
        'NativeBase.H1': {
            color: 'white',
        },
        'NativeBase.H2': {
            color: 'white',
        },
        'NativeBase.H3': {
            color: 'white',
        },
        'NativeBase.Title': {
            color: 'white',
        },
        'NativeBase.Subtitle': {
            color: 'white',
        },
    };

    const noBorder = {
        'NativeBase.Body': {
            borderBottomWidth: null,
        },
        'NativeBase.Right': {
            borderBottomWidth: null,
        },
        borderBottomWidth: null,
    };

    return {
        'NativeBase.InputGroup': {
            'NativeBase.Icon': {
                paddingRight: Scale(5),
            },
            'NativeBase.IconNB': {
                paddingRight: Scale(5),
            },
            'NativeBase.Input': {
                paddingHorizontal: Scale(5),
            },
            flex: 1,
            borderWidth: null,
            margin: -Scale(10),
            borderBottomColor: 'transparent',
        },
        '.searchBar': {
            'NativeBase.Item': {
                'NativeBase.Icon': {
                    backgroundColor: 'transparent',
                    color: variables.dropdownLinkColor,
                    fontSize:
                        platform === PLATFORM.IOS
                            ? variables.iconFontSize - Scale(10)
                            : variables.iconFontSize - Scale(5),
                    alignItems: 'center',
                    marginTop: Scale(2),
                    paddingRight: Spacing.spacer_2,
                },
                'NativeBase.IconNB': {
                    backgroundColor: 'transparent',
                    color: null,
                    alignSelf: 'center',
                },
                'NativeBase.Input': {
                    alignSelf: 'center',
                },
                alignSelf: 'center',
                alignItems: 'center',
                justifyContent: 'flex-start',
                flex: 1,
                height:
                    platform === PLATFORM.IOS ? Scale(30) : Spacing.spacer_7,
                borderColor: 'transparent',
                backgroundColor: '#fff',
                borderRadius: Scale(5),
            },
            'NativeBase.Button': {
                paddingHorizontal: platform === PLATFORM.IOS ? undefined : null,
                width: platform === PLATFORM.IOS ? undefined : 0,
                height: platform === PLATFORM.IOS ? undefined : 0,
            },
            backgroundColor: variables.toolbarInputColor,
            padding: Scale(10),
            marginLeft: null,
        },
        'NativeBase.CheckBox': {
            marginLeft: -Scale(10),
            marginRight: Scale(10),
        },
        '.first': {
            '.itemHeader': {
                paddingTop: variables.listItemPadding + Scale(3),
            },
        },
        '.itemHeader': {
            '.first': {
                paddingTop: variables.listItemPadding + Scale(3),
            },
            borderBottomWidth:
                platform === PLATFORM.IOS ? variables.borderWidth : null,
            marginLeft: null,
            padding: variables.listItemPadding,
            paddingLeft: variables.listItemPadding + Scale(5),
            paddingTop:
                platform === PLATFORM.IOS
                    ? variables.listItemPadding + Scale(25)
                    : undefined,
            paddingBottom:
                platform === PLATFORM.ANDROID
                    ? variables.listItemPadding + Scale(20)
                    : undefined,
            flexDirection: 'row',
            borderColor: variables.listBorderColor,
            'NativeBase.Text': {
                fontSize: Spacing.type_2,
                color:
                    platform === PLATFORM.IOS
                        ? undefined
                        : variables.listNoteColor,
            },
        },
        '.itemDivider': {
            borderBottomWidth: null,
            marginLeft: null,
            flexDirection: 'row',
            'NativeBase.Text': {
                marginBottom: 0,
                fontSize: Spacing.type_2,
            },
        },
        '.selected': {
            'NativeBase.Left': {
                ...selectedStyle,
            },
            'NativeBase.Body': {
                ...selectedStyle,
            },
            'NativeBase.Right': {
                ...selectedStyle,
            },
            ...selectedStyle,
        },
        'NativeBase.Left': {
            'NativeBase.Body': {
                marginLeft: Scale(10),
                alignItems: null,
                alignSelf: null,
            },
            'NativeBase.Icon': {
                width: variables.iconFontSize - Scale(10),
                fontSize: variables.iconFontSize - Scale(10),
            },
            'NativeBase.IconNB': {
                width: variables.iconFontSize - Scale(10),
                fontSize: variables.iconFontSize - Scale(10),
            },
            'NativeBase.Text': {
                alignSelf: 'center',
                marginBottom: 0,
            },
            'NativeBase.H3': {
                marginBottom: 0,
            },
            flexDirection: 'row',
        },
        'NativeBase.Body': {
            'NativeBase.Text': {
                marginRight: Spacing.spacer_4,
                marginBottom: 0,
            },
            'NativeBase.H3': {
                marginBottom: 0,
            },
            alignSelf: null,
            alignItems: null,
        },
        'NativeBase.Right': {
            'NativeBase.Badge': {
                alignSelf: null,
            },
            'NativeBase.PickerNB': {
                'NativeBase.Button': {
                    marginRight: -Scale(15),
                    'NativeBase.Text': {
                        color: variables.topTabBarActiveTextColor,
                    },
                },
            },
            'NativeBase.Button': {
                alignSelf: null,
                '.transparent': {
                    'NativeBase.Text': {
                        color: variables.topTabBarActiveTextColor,
                    },
                },
            },
            'NativeBase.Icon': {
                alignSelf: null,
                fontSize: variables.iconFontSize - Spacing.spacer_2,
                color: '#c9c8cd',
            },
            'NativeBase.IconNB': {
                alignSelf: null,
                fontSize: variables.iconFontSize - Spacing.spacer_2,
                color: '#c9c8cd',
            },
            'NativeBase.Text': {
                alignSelf: null,
                marginBottom: 0,
            },
            'NativeBase.H3': {
                marginBottom: 0,
            },
            'NativeBase.Thumbnail': {
                alignSelf: null,
            },
            'NativeBase.Image': {
                alignSelf: null,
            },
            'NativeBase.Radio': {
                alignSelf: null,
            },
            'NativeBase.Checkbox': {
                alignSelf: null,
            },
            'NativeBase.Switch': {
                alignSelf: null,
            },
            padding: null,
            flex: 0.28,
        },
        'NativeBase.Text': {
            alignSelf: 'center',
            marginBottom: 0,
        },
        '.last': {
            marginLeft: -(variables.listItemPadding + Scale(5)),
            paddingLeft: (variables.listItemPadding + Scale(5)) * 2,
            top: 1,
        },
        '.avatar': {
            'NativeBase.Left': {
                flex: 0,
                alignSelf: 'center',
                marginRight: Spacing.spacer_4,
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
                flex: 1,
                justifyContent: 'center',
            },
            'NativeBase.Right': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
                flex: 0,
                alignSelf: 'stretch',
                justifyContent: 'center',
            },
            '.noBorder': {
                ...noBorder,
            },
            paddingVertical: Spacing.spacer_2,
        },
        '.thumbnail': {
            'NativeBase.Left': {
                flex: 0,
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginLeft: null,
                },
                flex: 1,
                paddingVertical: variables.listItemPadding + Spacing.spacer_2,
                borderBottomWidth: variables.borderWidth,
                borderColor: variables.listBorderColor,
                marginLeft: variables.listItemPadding + Scale(5),
            },
            'NativeBase.Right': {
                'NativeBase.Button': {
                    '.transparent': {
                        'NativeBase.Text': {
                            fontSize: variables.listNoteSize,
                            color: variables.sTabBarActiveTextColor,
                        },
                    },
                    height: null,
                },
                flex: 0,
                justifyContent: 'center',
                alignSelf: 'stretch',
                paddingRight: variables.listItemPadding + Scale(5),
                paddingVertical: variables.listItemPadding + Scale(5),
                borderBottomWidth: variables.borderWidth,
                borderColor: variables.listBorderColor,
            },
            '.noBorder': {
                ...noBorder,
            },
            borderBottomWidth: null,
            paddingVertical: null,
            paddingRight: null,
        },
        '.icon': {
            '.last': {
                'NativeBase.Body': {
                    borderBottomWidth: null,
                },
                'NativeBase.Right': {
                    borderBottomWidth: null,
                },
                borderBottomWidth: variables.borderWidth,
                borderColor: variables.listBorderColor,
            },
            'NativeBase.Left': {
                'NativeBase.Button': {
                    'NativeBase.IconNB': {
                        marginHorizontal: null,
                        fontSize: variables.iconFontSize - Scale(5),
                    },
                    'NativeBase.Icon': {
                        marginHorizontal: null,
                        fontSize: variables.iconFontSize - Spacing.spacer_2,
                    },
                    alignSelf: 'center',
                    height: Scale(29),
                    width: Scale(29),
                    borderRadius: Scale(6),
                    paddingVertical: null,
                    paddingHorizontal: null,
                    alignItems: 'center',
                    justifyContent: 'center',
                },
                'NativeBase.Icon': {
                    width: variables.iconFontSize - Scale(5),
                    fontSize: variables.iconFontSize - Scale(2),
                },
                'NativeBase.IconNB': {
                    width: variables.iconFontSize - Scale(5),
                    fontSize: variables.iconFontSize - Scale(2),
                },
                paddingRight: variables.listItemPadding + Scale(5),
                flex: 0,
                height: Scale(44),
                justifyContent: 'center',
                alignItems: 'center',
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginLeft: null,
                    fontSize: Scale(17),
                },
                flex: 1,
                height: Scale(44),
                justifyContent: 'center',
                borderBottomWidth: 1 / PixelRatio.getPixelSizeForLayoutSize(1),
                borderColor: variables.listBorderColor,
            },
            'NativeBase.Right': {
                'NativeBase.Text': {
                    textAlign: 'center',
                    color: '#8F8E95',
                    fontSize: Scale(17),
                },
                'NativeBase.IconNB': {
                    color: '#C8C7CC',
                    fontSize: variables.iconFontSize - Scale(10),
                    alignSelf: 'center',
                    paddingLeft: Scale(10),
                    paddingTop: Scale(3),
                },
                'NativeBase.Icon': {
                    color: '#C8C7CC',
                    fontSize: variables.iconFontSize - Scale(10),
                    alignSelf: 'center',
                    paddingLeft: Scale(10),
                    paddingTop: Scale(3),
                },
                'NativeBase.Switch': {
                    marginRight:
                        Platform.OS === PLATFORM.IOS ? undefined : -Scale(5),
                    alignSelf: null,
                },
                'NativeBase.PickerNB': {
                    ...pickerTheme(),
                },
                flexDirection: 'row',
                alignItems: 'center',
                flex: 0,
                alignSelf: 'stretch',
                height: Scale(44),
                justifyContent: 'flex-end',
                borderBottomWidth: 1 / PixelRatio.getPixelSizeForLayoutSize(1),
                borderColor: variables.listBorderColor,
                paddingRight: variables.listItemPadding + Scale(5),
            },
            '.noBorder': {
                ...noBorder,
            },
            borderBottomWidth: null,
            paddingVertical: null,
            paddingRight: null,
            height: Scale(44),
            justifyContent: 'center',
        },
        '.noBorder': {
            ...noBorder,
        },
        '.header': {
            minHeight: getHeaderHeight,
            paddingTop:
                platform === PLATFORM.IOS
                    ? getStatusBarHeight
                    : Spacing.spacer_2,
            marginTop: platform === PLATFORM.IOS ? 0 : getStatusBarHeight,
            flexDirection: 'row',
            paddingHorizontal: Base,
            justifyContent: 'center',
            'NativeBase.Left': {
                flex: 1,
                'NativeBase.Button': {
                    marginLeft: Spacing.spacer_3,
                    marginRight: Spacing.spacer_3,
                },
            },
            'NativeBase.Body': {
                flex: 1,
                'NativeBase.Text': {
                    marginRight: 0,
                    textAlign: 'center',
                },
                'NativeBase.Button': {
                    marginLeft: Spacing.spacer_3,
                    marginRight: Spacing.spacer_3,
                },
            },
            'NativeBase.Right': {
                flex: 1,
                alignSelf: 'center',
                alignItems: 'flex-end',
                flexDirection: 'row',
                justifyContent: 'flex-end',
                'NativeBase.Button': {
                    marginLeft: Spacing.spacer_3,
                    marginRight: Spacing.spacer_3,
                },
            },
        },
        '.white': {
            'NativeBase.Left': {
                ...textWhite,
            },
            'NativeBase.Body': {
                ...textWhite,
            },
            'NativeBase.Right': {
                ...textWhite,
            },
        },
        '.ammounts': {
            'NativeBase.Left': {
                flex: 1,
                flexGrow: 1,
                alignSelf: 'center',
                marginRight: Spacing.spacer_4,
            },
            'NativeBase.Body': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
                flex: 1,
                justifyContent: 'center',
            },
            'NativeBase.Right': {
                'NativeBase.Text': {
                    marginBottom: 0,
                },
                flex: 0,
                alignSelf: 'stretch',
                justifyContent: 'center',
            },
            '.noBorder': {
                ...noBorder,
            },
            paddingVertical: Spacing.spacer_2,
        },
        '.small': {
            paddingTop: Spacing.spacer_1,
            paddingBottom: Spacing.spacer_1,
        },
        alignItems: 'center',
        flexDirection: 'row',
        paddingVertical: Spacing.spacer_2,
        paddingHorizontal: Spacing.spacer_4,
        borderBottomWidth: 1 / PixelRatio.getPixelSizeForLayoutSize(1),
        backgroundColor: variables.listBg,
        borderColor: variables.listBorderColor,
    };
};
