// react-native-minerva/native-base/components/Segment.js
// @flow
import { Spacing, HelperStyle, Fonts, Scale } from '../../src/index';

export default () => ({
    flexDirection: 'row',
    justifyContent: 'center',
    borderWidth: 0,
    borderColor: 'transparent',
    width: '100%',
    backgroundColor: 'transparent',
    height: 'auto',
    overflow: 'visible',
    marginBottom: Spacing.spacer_1,
    'NativeBase.Button': {
        height: Spacing.spacer_7,
        alignSelf: 'flex-start',
        paddingTop: 0,
        paddingBottom: 0,
        flexGrow: 1,
        overflow: 'visible',
        backgroundColor: HelperStyle.bg_white.backgroundColor,
        borderColor: 'transparent',
        shadowColor: HelperStyle.gray_light.color,
        shadowOffset: { width: 0, height: 3 },
        shadowOpacity: 0.2,
        shadowRadius: Spacing.spacer_1,
        borderRadius: 0,
        elevation: 2,
        textTransform: 'none',
        '.active': {
            backgroundColor: HelperStyle.primary_light.color,
            'NativeBase.Text': {
                color: HelperStyle.white.color,
            },
            'NativeBase.Icon': {
                color: HelperStyle.white.color,
            },
        },
        '.first': {
            borderTopLeftRadius: Spacing.spacer_4,
            borderBottomLeftRadius: Spacing.spacer_4,
        },
        '.last': {
            borderTopRightRadius: Spacing.spacer_4,
            borderBottomRightRadius: Spacing.spacer_4,
        },
        'NativeBase.Text': {
            fontFamily: Fonts.fontFamily,
            textTransform: 'none',
            color: HelperStyle.gray.color,
            fontSize: Spacing.type_2,
        },
        'NativeBase.Icon': {
            fontSize: Scale(22),
            paddingTop: 0,
            color: HelperStyle.gray.color,
        },
    },
});
