// react-native-minerva/native-base/components/Separator.js
// @flow

import { Spacing, Scale } from '../../src/index';
import variable from '../variables/platform';

export default (variables /* : * */ = variable) => ({
    '.group': {
        height: Scale(50),
        paddingVertical: variables.listItemPadding - Spacing.spacer_2,
        paddingTop: variables.listItemPadding + Spacing.spacer_3,
        '.bordered': {
            height: Scale(50),
            paddingVertical: variables.listItemPadding - Spacing.spacer_2,
            paddingTop: variables.listItemPadding + Spacing.spacer_3,
        },
    },
    '.bordered': {
        '.noTopBorder': {
            borderTopWidth: 0,
        },
        '.noBottomBorder': {
            borderBottomWidth: 0,
        },
        height: Scale(35),
        paddingTop: variables.listItemPadding + Scale(2),
        paddingBottom: variables.listItemPadding,
        borderBottomWidth: variables.borderWidth,
        borderTopWidth: variables.borderWidth,
        borderColor: variables.listBorderColor,
    },
    'NativeBase.Text': {
        fontSize: variables.tabBarTextSize - Scale(2),
        color: '#777',
    },
    '.noTopBorder': {
        borderTopWidth: 0,
    },
    '.noBottomBorder': {
        borderBottomWidth: 0,
    },
    height: Scale(38),
    backgroundColor: '#F0EFF5',
    flex: 1,
    justifyContent: 'center',
    paddingLeft: variables.listItemPadding + Scale(5),
});
