// react-native-minerva/native-base/components/TabBar.js
import { Scale, minerva } from '../../src/index';

export default (variables /* : * */ = minerva) => ({
    '.tabIcon': {
        height: undefined,
    },
    '.vertical': {
        height: Scale(60),
    },
    'NativeBase.Button': {
        '.transparent': {
            'NativeBase.Text': {
                fontSize: variables.tabFontSize,
                color: variables.sTabBarActiveTextColor,
                fontFamily: variables.fontFamily,
            },
            'NativeBase.IconNB': {
                color: variables.sTabBarActiveTextColor,
            },
        },
        'NativeBase.IconNB': {
            color: variables.sTabBarActiveTextColor,
        },
        'NativeBase.Text': {
            fontSize: variables.tabFontSize,
            color: variables.sTabBarActiveTextColor,
            fontFamily: variables.fontFamily,
        },
        '.isTabActive': {
            'NativeBase.Text': {
                fontFamily: variables.fontFamilyBold,
            },
        },
        flex: 1,
        alignSelf: 'center',
        alignItems: 'center',
        justifyContent: 'center',
        borderRadius: null,
        borderBottomColor: 'transparent',
        backgroundColor: variables.tabBgColor,
    },
    height: null,
    flexDirection: 'row',
    justifyContent: 'center',
    alignSelf: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderTopWidth: 0,
    borderLeftWidth: 0,
    borderRightWidth: 0,
    borderBottomColor: '#ccc',
    backgroundColor: variables.tabBgColor,
});
