// react-native-minerva/native-base/components/TabHeading.js
// @flow

import { Spacing, minerva, Fonts, Scale } from '../../src/index';
import { PLATFORM } from '../variables/minerva';

export default (variables /* : * */ = minerva) => {
    const { platform } = variables;

    return {
        flex: 1,
        flexDirection: 'column',
        backgroundColor: 'transparent',
        justifyContent: 'center',
        alignContent: 'center',
        paddingHorizontal: Spacing.spacer_4,
        paddingVertical: Spacing.spacer_3,
        '.scrollable': {
            paddingHorizontal: Spacing.spacer_4,
            paddingVertical: Spacing.spacer_3,
            flex: platform === PLATFORM.ANDROID ? 0 : 1,
            minWidth: platform === PLATFORM.ANDROID ? undefined : Scale(60),
        },
        'NativeBase.Text': {
            marginBottom: 0,
            alignSelf: 'center',
            textAlign: 'center',
        },
        'NativeBase.Text.smaller': {
            marginBottom: 0,
            lineHeight: Spacing.lineHeight_1,
        },
        'NativeBase.Icon': {
            color: variables.gray_light,
            fontSize: platform === PLATFORM.IOS ? Spacing.type_6 : undefined,
        },
        '.active': {
            alignItems: 'center',
            'NativeBase.Text': {
                color: variables.primary,
            },
            'NativeBase.Text.smaller': {
                color: variables.gray,
            },
            'NativeBase.Icon': {
                color: variables.primary,
            },
        },
    };
};
