// react-native-minerva/native-base/components/Toast.js
// @flow

import variable from '../variables/platform';
import { Spacing, Scale } from '../../src/index';
import { PLATFORM } from '../variables/commonColor';

export default (variables /* : * */ = variable) => {
    const { platform } = variables;

    return {
        '.danger': {
            backgroundColor: variables.brandDanger,
        },
        '.warning': {
            backgroundColor: variables.brandWarning,
        },
        '.success': {
            backgroundColor: variables.brandSuccess,
        },
        backgroundColor: 'rgba(0,0,0,0.8)',
        borderRadius: platform === PLATFORM.IOS ? Scale(5) : 0,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        padding: Scale(10),
        minHeight: Scale(50),
        'NativeBase.Text': {
            color: '#fff',
            flex: 1,
        },
        'NativeBase.Button': {
            backgroundColor: 'transparent',
            height: Scale(30),
            elevation: 0,
            'NativeBase.Text': {
                fontSize: Spacing.type_2,
            },
        },
    };
};
