// react-native-minerva/src/components/chip/style.js
import { StyleSheet } from 'react-native';
import { Scale } from '../../index';

const Style = (HelperStyle) =>
    StyleSheet.create({
        defaultStyle: {
            ...HelperStyle.flx_row,
            ...HelperStyle.bw_1,
            ...HelperStyle.bc_primary_light,
            ...HelperStyle.bg_white,
            minHeight: Scale(32),
            ...HelperStyle.br_3,
            ...HelperStyle.ai_c,
            ...HelperStyle.as_fs,
            ...HelperStyle.jc_c,
        },
        defaultStyleText: {
            ...HelperStyle.ml_4,
            ...HelperStyle.mr_4,
            ...HelperStyle.ff_fontFamily,
            ...HelperStyle.mb_0,
            ...HelperStyle.primary,
        },
        activeButton: {
            ...HelperStyle.bg_primary_light,
        },
        activeText: {
            ...HelperStyle.white,
        },
        big: {
            height: Scale(56),
            ...HelperStyle.ph_2,
        },
        bigIcon: {
            ...HelperStyle.pv_4,
            ...HelperStyle.pl_3,
        },
        bigger: {
            ...HelperStyle.flx_col,
            ...HelperStyle.pa_5,
        },
        buttonIcon: {
            ...HelperStyle.flx_row,
        },
        icon: {
            ...HelperStyle.fs_3,
            ...HelperStyle.primary,
            ...HelperStyle.ml_3,
        },
        removable: {
            ...HelperStyle.fs_3,
            ...HelperStyle.gray,
            ...HelperStyle.pr_2,
        },
        textRemovable: {
            ...HelperStyle.mr_2,
        },
        iconBigger: {
            ...HelperStyle.fs_7,
            ...HelperStyle.ml_0,
            ...HelperStyle.ma_0,
        },
        textIcon: {
            ...HelperStyle.ml_2,
        },
        textAvatar: {
            ...HelperStyle.ml_2,
            ...HelperStyle.mr_2,
        },
        textBigger: {
            ...HelperStyle.ml_0,
            ...HelperStyle.mr_0,
            ...HelperStyle.mt_3,
        },
        avatar: {
            minWidth: Scale(25),
            height: Scale(25),
            width: Scale(25),
            ...HelperStyle.ml_1,
            ...HelperStyle.pl_0,
            ...HelperStyle.pr_0,
        },
        disabled: {
            opacity: 0.5,
        },
        textBadge: {
            ...HelperStyle.ml_2,
            ...HelperStyle.mr_4,
        },
    });

export default Style;
