// react-native-minerva/src/components/chip/type.js
import PropTypes from 'prop-types';

const chipPropTypes = {
    /**
     * Boleano para definit estilos para estado seleccionado.
     */
    selected: PropTypes.bool,
    /**
     * Texto del <Chip />.
     */
    title: PropTypes.string.isRequired,
    /**
     * Boleano para definir tamaño Big.
     */
    big: PropTypes.bool,
    /**
     * Boleano para definir tamaño bigger.
     */
    bigger: PropTypes.bool,
    /**
     * Boleano para mostrar <Icon /> remover.
     */
    removable: PropTypes.bool,
    /**
     * Nombre de icono. Si esta definido se muestra en el lado izquierdo.
     */
    icon: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Boleano para mostrar componente <Avatar /> en el lado izquierdo.
     */
    avatar: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Acción del <Chip />.
     */
    onPress: PropTypes.func,
    /**
     * Propiedades para configurar componente <Icon />.
     */
    iconProps: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
        PropTypes.bool,
    ]),
    /**
     * Propiedades para configurar componente <Avatar />.
     */
    avatarProps: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
        PropTypes.bool,
    ]),
    /**
     * Propiedades para el title dentro del <Chip />.
     */
    textProps: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
        PropTypes.bool,
    ]),
    /**
     * Propiedades para el <Icon /> remover.
     */
    removableProps: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
        PropTypes.bool,
    ]),
    /**
     * Boleano para habilitar o deshabilitar.
     */
    disabled: PropTypes.bool,
    /**
     * Boleano para ocultar/mostrar.
     */
    visible: PropTypes.bool,
    /**
     * Estilos para <Chip />.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Boleano para mostrar componente <Badge /> en el lado izquierdo.
     */
    badge: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Propiedades para configurar componente <Badge />.
     */
    badgeProps: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
        PropTypes.bool,
    ]),
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Identificador para accesibilidad.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad.
     */
    accessibilityHint: PropTypes.string,
};

const chipDefaultProps = {
    selected: false,
    big: false,
    bigger: false,
    icon: false,
    avatar: false,
    removable: false,
    onPress: () => null,
    iconProps: {},
    avatarProps: {},
    textProps: {},
    removableProps: {},
    visible: true,
    disabled: false,
    style: {},
    badge: false,
    badgeProps: {},
    accessibilityLabel: '',
    accessibilityHint: '',
    testID: '',
};

export { chipPropTypes, chipDefaultProps };
