// react-native-minerva/src/components/itemMovimiento/ItemMovimientoMedia.js
import React from 'react';
import { Image } from 'react-native';
import { HelperStyle, Badge, Icon } from '../../index';
import { propTypesMedia, defaultPropsMedia } from './type';
import ITEMMOVIMIENTO_STR from './itemMovimientoStr';

/**
 * Componente para mostrar una imagen o icono en el movimiento.
 *
 * @component
 */
const ItemMovimientoMedia = ({
    iconName,
    iconStyle,
    shape,
    imageSource,
    testID,
    accessibilityLabel,
    accessibilityHint,
    style,
}) => {
    const styleShape = {
        square: HelperStyle.br_2,
        circle: HelperStyle.borderRadius(ITEMMOVIMIENTO_STR.mediaSize),
    };
    const imageMovimiento = () => (
        <Image
            style={[
                HelperStyle.width(ITEMMOVIMIENTO_STR.mediaSize),
                HelperStyle.height(ITEMMOVIMIENTO_STR.mediaSize),
                HelperStyle.bg_gray_background,
                styleShape[shape],
                style,
            ]}
            source={imageSource}
            testID={testID}
            accessibilityLabel={accessibilityLabel}
            accessibilityHint={accessibilityHint}
        />
    );
    const iconMovimiento = () => (
        <Badge
            smaller
            size={ITEMMOVIMIENTO_STR.mediaSize}
            style={[HelperStyle.br_2, styleShape[shape], style]}
        >
            <Icon
                name={iconName}
                style={[HelperStyle.fs_1, HelperStyle.gray, iconStyle]}
                testID={testID}
                accessibilityLabel={accessibilityLabel}
                accessibilityHint={accessibilityHint}
            />
        </Badge>
    );
    const renderType = () => {
        if (imageSource) return imageMovimiento();
        return iconMovimiento();
    };
    return renderType();
};

ItemMovimientoMedia.propTypes = {
    iconName: propTypesMedia.iconName,
    shape: propTypesMedia.shape,
    imageSource: propTypesMedia.imageSource,
    accessibilityLabel: propTypesMedia.accessibilityLabel,
    testID: propTypesMedia.testID,
    accessibilityHint: propTypesMedia.accessibilityHint,
};
ItemMovimientoMedia.defaultProps = {
    iconName: defaultPropsMedia.iconName,
    shape: defaultPropsMedia.shape,
    imageSource: defaultPropsMedia.imageSource,
    accessibilityLabel: defaultPropsMedia.accessibilityLabel,
    accessibilityHint: defaultPropsMedia.accessibilityHint,
    testID: defaultPropsMedia.testID,
};

export default ItemMovimientoMedia;
