// react-native-minerva/src/components/input/style.js
import { StyleSheet } from 'react-native';
import { minerva, Spacing, Fonts, Scale, Platform } from '../../index';

const { inputHeightBase } = minerva();

const Style = ({ maxHeight, HelperStyle, isDisabled }) =>
    StyleSheet.create({
        itemMultiline: {
            height: 'auto',
            ...(Platform.android && HelperStyle.pb_0),
        },
        inputMultiline: {
            height: 'auto',
            maxHeight: Scale(maxHeight),
            ...HelperStyle.pt_0,
            ...HelperStyle.lh_2,
        },
        LabelContainer: {
            position: 'absolute',
            left: Scale(10),
            zIndex: 2,
            paddingHorizontal: Scale(6),
            top: Scale(15),
        },
        contentLeft: {
            justifyContent: 'center',
            alignItems: 'center',
            paddingRight: Spacing.spacer_3,
        },
        contentRight: {
            paddingLeft: Spacing.spacer_3,
        },
        label: {
            fontSize: Spacing.type_3,
            zIndex: 2,
            fontFamily: Fonts.fontFamily,
            lineHeight: Spacing.lineHeight_2,
        },
        inputHeightBase: {
            height: inputHeightBase,
        },
        inputHeightBaseMultiline: {
            height: null,
        },
        clearPadding: {
            paddingRight: Spacing.spacer_0,
            paddingLeft: Spacing.spacer_0,
            paddingTop: Spacing.spacer_0,
            paddingBottom: Spacing.spacer_0,
        },
        disabledMask: {
            opacity: isDisabled ? 0.4 : 1,
        },
        centerSideWithError: {
            marginBottom: Scale(28.5),
        },
    });

export default Style;
