// react-native-minerva/src/hook/useSafeAreaInsetsStyle.js
import { useSafeAreaInsets } from 'react-native-safe-area-context';

/**
 * Wrapper del hook useSafeAreaInsets para obtener estilos con las dimensiones del area segura.
 *
 * Condiciones para ser usada:
 * 1. Debe ser usada dentro del cuerpo de un componente funcional.
 * 2. El componente funcional debe ser hijo o estar dentro de un safe-area-context.
 *
 * @returns {object} - Objecto con estilos con las dimensiones del area segura.
 */
const useSafeAreaInsetsStyle = () => {
    const insets = useSafeAreaInsets();
    return {
        pv: {
            paddingBottom: Math.max(insets.bottom, 16),
            paddingTop: Math.max(insets.top, 16),
        },
        pb: { paddingBottom: Math.max(insets.bottom, 16) },
        pt: { paddingTop: Math.max(insets.top, 16) },
        mv: {
            marginBottom: Math.max(insets.bottom, 16),
            marginTop: Math.max(insets.top, 16),
        },
        mt: { marginTop: Math.max(insets.top, 16) },
        mb: { marginBottom: Math.max(insets.bottom, 16) },
    };
};

export default useSafeAreaInsetsStyle;
