// react-native-minerva/src/components/selloComprobante/selloComprobante.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Assets from 'bch-assets';
import { Image } from 'react-native';
import Style from './style';
import SELLOCOMPROBANTE_STR from './selloComprobanteStr';

/**
 * Componente para mostrar sello comprobante por marca y app.
 *
 * @component
 */
const SelloComprobante = ({ type }) => {
    const configSellos = {
        appMiBancoBCH: Assets.sello.miBanco.bch,
        appMiBancoBEC: Assets.sello.miBanco.bec,
        appMiBancoCCH: Assets.sello.miBanco.cch,
        appMiPago: Assets.sello.miPago.miPago,
        appMiPagoPaga2: Assets.sello.miPago.paga2,
        appMiInversion: Assets.sello.miInversion,
        appBanconexion: Assets.sello.banconexion,
        appMiSeguro: Assets.sello.miSeguro,
    };
    return <Image style={Style.selloComprobante} source={configSellos[type]} />;
};

SelloComprobante.propTypes = {
    /**
     * Tipo de sello.
     */
    type: PropTypes.oneOf(Object.values(SELLOCOMPROBANTE_STR.type)),
};
SelloComprobante.defaultProps = {
    type: SELLOCOMPROBANTE_STR.type.appMiBancoBCH,
};

export default memo(SelloComprobante);
