// react-native-minerva/src/components/squareButtonsGroup/style.js
import { StyleSheet } from 'react-native';
import { Scale } from '../../index';

const Style = (HelperStyle, centered, insideText, selected) =>
    StyleSheet.create({
        list: {
            ...HelperStyle.flx_row,
            ...HelperStyle.flx_wrap,
            ...HelperStyle.jc_c,
            ...HelperStyle.pa_2,
        },
        item: {
            ...HelperStyle.w_50,
            ...(centered && HelperStyle.as_c),
        },
        itemStyle: {
            ...HelperStyle.ml_2,
            ...HelperStyle.mr_2,
            minHeight: Scale(80),
            ...(selected && HelperStyle.bc_primary),
            ...(!insideText && HelperStyle.mb_2),
        },
        bodyCard: {
            ...HelperStyle.jc_c,
            ...HelperStyle.ai_c,
            ...(insideText ? HelperStyle.pt_2 : HelperStyle.pv_2),
        },
        iconStyle: {
            ...HelperStyle.ta_c,
            ...(insideText ? HelperStyle.fs_5 : HelperStyle.fs_7),
        },
        textInside: {
            ...HelperStyle.mb_0,
            ...HelperStyle.lh_none,
            ...HelperStyle.mt_3,
        },
        textOutside: {
            ...HelperStyle.ta_c,
            ...HelperStyle.lh_none,
        },
    });

export default Style;
