// react-native-minerva/src/components/carousel/style.js
import { StyleSheet } from 'react-native';
import { Scale, Spacing } from '../../index';

const circleDefault = {
    alignSelf: 'center',
    position: 'absolute',
};

const style = StyleSheet.create({
    circleSmall: {
        ...circleDefault,
        top: 0,
        bottom: 0,
        left: Spacing.spacer_7,
        height: Spacing.spacer_6,
        width: Spacing.spacer_6,
        borderRadius: Spacing.spacer_6 / 2,
    },
    circleMedium: {
        ...circleDefault,
        top: Scale(10),
        right: 0,
        height: Scale(72),
        width: Scale(72),
        borderRadius: Scale(72) / 2,
    },
    circleLarge: {
        ...circleDefault,
        left: 0,
        bottom: 0,
        right: 0,
        height: Scale(120),
        width: Scale(120),
        borderRadius: Scale(120) / 2,
    },
    circleContainer: {
        position: 'absolute',
        width: Scale(200),
        alignSelf: 'center',
        height: Scale(160),
        zIndex: 1,
        transform: [{ translateY: -Scale(100) }],
    },
    carouselContainer: {
        zIndex: 2,
    },
});

export default style;
