// react-native-minerva/src/util/helpers/styles/flexbox.js
const valueNames = {
    fs: 'flex-start',
    fe: 'flex-end',
    s: 'stretch',
    c: 'center',
    sb: 'space-between',
    sa: 'space-around',
};

export default {
    flx_0: {
        flex: 0,
    },
    flx_i: {
        flex: 1,
    },
    flx_2: {
        flex: 2,
    },
    flx_3: {
        flex: 3,
    },
    flx_grow: {
        flexGrow: 1,
    },
    flx_grow_0: {
        flexGrow: 0,
    },
    flx_shrink: {
        flexShrink: 1,
    },
    flx_shrink_0: {
        flexShrink: 0,
    },
    flx_row: {
        flexDirection: 'row',
    },
    'flx_row-reverse': {
        flexDirection: 'row-reverse',
    },
    flx_col: {
        flexDirection: 'column',
    },
    'flx_col-reverse': {
        flexDirection: 'column-reverse',
    },
    flx_wrap: {
        flexWrap: 'wrap',
    },
    ac_fs: {
        alignContent: valueNames.fs,
    },
    ac_fe: {
        alignContent: valueNames.fe,
    },
    ac_s: {
        alignContent: valueNames.s,
    },
    ac_c: {
        alignContent: valueNames.c,
    },
    ac_sb: {
        alignContent: valueNames.sb,
    },
    ac_sa: {
        alignContent: valueNames.sa,
    },
    ai_fs: {
        alignItems: valueNames.fs,
    },
    ai_c: {
        alignItems: valueNames.c,
    },
    ai_fe: {
        alignItems: valueNames.fe,
    },
    as_fs: {
        alignSelf: valueNames.fs,
    },
    as_c: {
        alignSelf: valueNames.c,
    },
    as_fe: {
        alignSelf: valueNames.fe,
    },
    as_s: {
        alignSelf: valueNames.s,
    },
    as_a: {
        alignSelf: 'auto',
    },
    jc_fs: {
        justifyContent: valueNames.fs,
    },
    jc_fe: {
        justifyContent: valueNames.fe,
    },
    jc_c: {
        justifyContent: valueNames.c,
    },
    jc_sb: {
        justifyContent: valueNames.sb,
    },
    jc_sa: {
        justifyContent: valueNames.sa,
    },
};
