// react-native-minerva/src/util/helpers/styles/spacing.js
import { toPairs, forEach } from 'lodash';
import { Platform } from '../../../constants/index';
import {
    getStatusBarHeight,
    getHeaderHeight,
    getHeaderStatusBarHeight,
} from '../../index';

export default (REM_SCALE) => {
    const what = toPairs({
        m: 'margin',
        p: 'padding',
    });
    const where = toPairs({
        a: '',
        h: 'Horizontal',
        v: 'Vertical',
        t: 'Top',
        r: 'Right',
        b: 'Bottom',
        l: 'Left',
    });
    const spacing = {};
    forEach(what, ([whatShort, whatLong]) => {
        forEach(where, ([whereShort, whereLong]) => {
            forEach(REM_SCALE, (scale, idx) => {
                spacing[`${whatShort}${whereShort}${idx}`] = {
                    [`${whatLong}${whereLong}`]: scale,
                };
            });
        });
    });

    const spacingHeader = {
        mt_header: {
            marginTop: getHeaderHeight,
        },
        pt_header: {
            paddingTop: getHeaderHeight,
        },
        mt_statusbar: {
            marginTop: getStatusBarHeight,
        },
        pt_statusbar: {
            paddingTop: getStatusBarHeight,
        },
        mt_headerStatusbar: {
            marginTop: getHeaderStatusBarHeight,
        },
        pt_headerStatusbar: {
            paddingTop: getHeaderStatusBarHeight,
        },
        /*
         * Header por plataforma:
         * IOS: Altura header mas statusbar
         * Android: Altura header
         */
        mt_headerPlatform: {
            marginTop: Platform.ios
                ? getHeaderStatusBarHeight
                : getHeaderHeight,
        },
        pt_headerPlatform: {
            paddingTop: Platform.ios
                ? getHeaderStatusBarHeight
                : getHeaderHeight,
        },
    };

    return {
        ...spacing,
        ...spacingHeader,
    };
};
