// react-native-minerva/src/components/cardTrampolin/components/ImageMask.js
import React, { memo } from 'react';
import { Image } from 'react-native';
import Images from 'bch-assets';
import MaskedView from '@react-native-community/masked-view';
import { HelperStyle } from '../../../index';
import { propTypesImageMask, defaultPropsImageMask } from '../types';
import CARDTRAMPOLIN_STR from '../CardTrampolinStr';

const ImageMask = ({
    type,
    image,
    styleImageMask,
    imageTestID,
    imageAccessibilityLabel,
    imageAccessibilityHint,
}) => {
    const {
        components: {
            cardTrampolin: { img, mask, maskMiInversion, men },
        },
    } = Images;

    const defaultMaskedViewProps = {
        waves_left: {
            sourceImage: img,
            sourceImageMask: mask,
        },
        miinversion: {
            sourceImage: men,
            sourceImageMask: maskMiInversion,
        },
    };

    const renderImage = (source) => (
        <Image
            style={[HelperStyle.w_100, HelperStyle.h_100, styleImageMask]}
            source={source}
            accessible
            testID={imageTestID}
            accessibilityRole="image"
            accessibilityLabel={imageAccessibilityLabel}
            accessibilityHint={imageAccessibilityHint}
        />
    );

    return (
        <MaskedView
            testID={CARDTRAMPOLIN_STR.maskedView.testID}
            style={[
                HelperStyle.flx_i,
                HelperStyle.flx_row,
                HelperStyle.w_100,
                HelperStyle.h_100,
            ]}
            maskElement={renderImage(
                defaultMaskedViewProps[type]?.sourceImageMask,
            )}
        >
            {renderImage(image || defaultMaskedViewProps[type]?.sourceImage)}
        </MaskedView>
    );
};

ImageMask.propTypes = propTypesImageMask;
ImageMask.defaultProps = defaultPropsImageMask;

export default memo(ImageMask);
