// react-native-minerva/src/components/button/style.js
import { StyleSheet } from 'react-native';
import { Spacing, Scale, ColorsBCH } from '../../index';

const style = StyleSheet.create({
    linearGradient: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
        borderColor: ColorsBCH.transparent,
        borderRadius: Scale(57),
    },
    buttonHeightSmaller: {
        height: Spacing.spacer_6,
    },
    buttonHeightSmall: {
        height: Spacing.spacer_7,
    },
    buttonHeightMedium: {
        height: Scale(48),
    },
    buttonHeightTiny: {
        height: Scale(24),
    },
    buttonHeightDefault: {
        height: Scale(56),
    },
    buttonDefaultPadding: {
        paddingHorizontal: Spacing.spacer_5,
        paddingVertical: Spacing.spacer_4,
    },
    buttonSmallPadding: {
        paddingHorizontal: Spacing.spacer_4,
        paddingVertical: Spacing.spacer_2,
    },
    buttonSmallerPadding: {
        paddingHorizontal: Spacing.spacer_4,
        paddingVertical: Spacing.spacer_1,
    },
    buttonTinyPadding: {
        paddingHorizontal: Spacing.spacer_3,
        paddingVertical: Spacing.spacer_1,
    },
    // padding icon
    buttonIconBigPadding: {
        paddingHorizontal: Spacing.spacer_3,
        paddingVertical: Spacing.spacer_3,
    },
    buttonIconDefaultPadding: {
        paddingHorizontal: Spacing.spacer_4,
        paddingVertical: Spacing.spacer_4,
    },
    buttonIconSmallPadding: {
        paddingHorizontal: Spacing.spacer_3,
        paddingVertical: Spacing.spacer_3,
    },
    buttonIconSmallerPadding: {
        paddingHorizontal: Spacing.spacer_2,
        paddingVertical: Spacing.spacer_2,
    },
    buttonIconTinyPadding: {
        paddingHorizontal: Spacing.spacer_2,
        paddingVertical: Spacing.spacer_2,
    },
    buttonIconMediumPadding: {
        paddingHorizontal: Spacing.spacer_3,
        paddingVertical: Spacing.spacer_2,
    },
    // padding icon sin border (-2 ancho del border)
    buttonIconBigPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_3 - 2,
        paddingVertical: Spacing.spacer_3 - 2,
    },
    buttonIconDefaultPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_4 - 2,
        paddingVertical: Spacing.spacer_4 - 2,
    },
    buttonIconSmallPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_3 - 2,
        paddingVertical: Spacing.spacer_3 - 2,
    },
    buttonIconSmallerPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_2 - 2,
        paddingVertical: Spacing.spacer_2 - 2,
    },
    buttonIconTinyPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_2 - 2,
        paddingVertical: Spacing.spacer_2 - 2,
    },
    buttonIconMediumPaddingNoBorder: {
        paddingHorizontal: Spacing.spacer_4 - 2,
        paddingVertical: Spacing.spacer_2 - 2,
    },
    buttonWidthFluid: {
        flex: null,
        flexDirection: null,
        justifyContent: 'center',
        alignItems: null,
        width: 'auto',
    },
    rippleStyle: {
        flex: 1,
        borderRadius: Scale(57),
    },
    borderFooterTab: {
        height: Spacing.spacer_1,
        borderRadius: Spacing.radius_3,
        position: 'absolute',
        width: '100%',
        top: 0,
        left: 0,
        zIndex: 100,
    },
    buttonContentIconText: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: ColorsBCH.transparent,
        zIndex: 2,
    },
});

export default style;
