// react-native-minerva/src/components/cardItemProducto/style.js
import { StyleSheet } from 'react-native';
import { Spacing, Scale } from '../../index';

const style = StyleSheet.create({
    icon: {
        fontSize: Scale(35),
    },
    contentRightButton: {
        flex: 0.7,
        marginLeft: Spacing.spacer_1,
    },
    contentRightButtonCollapse: {
        maxWidth: Spacing.spacer_5,
        marginLeft: Spacing.spacer_1,
    },
    contentLeftButton: {
        flex: 0.2,
    },
    styleLogoMastercard: {
        width: Scale(48),
        height: Spacing.spacer_6,
        alignSelf: 'center',
        resizeMode: 'contain',
    },
    heightTitleCard: {
        height: Spacing.spacer_6,
    },
    cardItemHeader: {
        paddingTop: 0,
        paddingRight: 0,
        paddingBottom: 0,
        paddingLeft: 0,
        alignItems: 'stretch',
    },
    cardItemHeaderTitleContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
    },
    cardItemHeaderTitle: {
        marginBottom: 0,
        flex: 1,
        flexWrap: 'wrap',
    },
    LinearGradientStyle: {
        flex: 1,
        flexDirection: 'row',
        borderTopLeftRadius: Spacing.radius_3,
        borderTopRightRadius: Spacing.radius_3,
    },
    imageBackgroundTop: {
        width: '100%',
        height: '100%',
        position: 'absolute',
        top: 0,
        left: 0,
        zIndex: 0,
        resizeMode: 'contain',
    },
    minHeightCard: {
        minHeight: Scale(176),
    },
    noBorderRadiusTop: {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
    },
    headerContent: {
        flex: 1,
        paddingTop: Spacing.spacer_4,
        paddingHorizontal: Spacing.spacer_4,
        paddingBottom: Spacing.spacer_4,
    },
    lineSeparatorStyle: {
        borderWidth: 0.5,
    },
    fontSizeIconStyle: {
        fontSize: Spacing.type_7,
        maxWidth: Scale(32),
        maxHeight: Scale(32),
    },
});

export default style;
