// react-native-minerva/src/components/selects/components/DateTimePicker.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import DateTimePicker from '@react-native-community/datetimepicker';
import { View, Platform, HelperStyle } from '../../../index';

const DateTimePickerBCH = (props) => {
    const { style, iosMode, androidMode, themeVariant } = props;
    const modePlatform = Platform.ios ? iosMode : androidMode;
    const paddingBottom =
        Platform.ios && Platform.version < 15 ? HelperStyle.pb_4 : null;
    return (
        <View style={[paddingBottom, style]}>
            <DateTimePicker
                {...props}
                mode={modePlatform}
                themeVariant={themeVariant}
            />
        </View>
    );
};

DateTimePickerBCH.propTypes = {
    is24Hour: PropTypes.bool,
    display: PropTypes.string,
    androidMode: PropTypes.oneOf(['date', 'time']),
    iosMode: PropTypes.oneOf(['date', 'time', 'datetime', 'countdown']),
    themeVariant: PropTypes.oneOf(['dark', 'light']),
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};
DateTimePickerBCH.defaultProps = {
    is24Hour: true,
    display: Platform.ios ? 'inline' : 'default',
    androidMode: 'date',
    iosMode: 'date',
    themeVariant: 'light',
    style: {},
};

export default memo(DateTimePickerBCH);
