// react-native-minerva/src/components/headerHomeApp/HeaderHomeApp.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Header, Left, Body, Right, Validate } from '../../index';

/**
 * Componente para mostrar encabezado Home.
 *
 * @component
 */
const HeaderHomeApp = ({
    renderLeft,
    renderBody,
    renderRight,
    disabled,
    accessibilityLabel,
    testID,
    accessibilityHint,
}) => (
    <Header
        disabled={disabled}
        accessibilityLabel={accessibilityLabel}
        accessibilityHint={accessibilityHint}
        testID={testID}
        accessible={false}
    >
        <Left>{Validate.render(renderLeft)}</Left>
        <Body>{Validate.render(renderBody)}</Body>
        <Right>{Validate.render(renderRight)}</Right>
    </Header>
);

HeaderHomeApp.propTypes = {
    /**
     * Renderiza un componente en el lado izquierdo.
     */
    renderLeft: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /**
     * Renderiza un componente en el centro.
     */
    renderBody: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /**
     * Renderiza un componente en el lado derecho.
     */
    renderRight: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    /**
     * Boleano para habilitar/desahabilitar.
     */
    disabled: PropTypes.bool,
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Identificador para accesibilidad.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Accesibilidad: ayuda a los usuarios a comprender qué sucederá cuando realicen una acción en el elemento.
     */
    accessibilityHint: PropTypes.string,
};
HeaderHomeApp.defaultProps = {
    renderLeft: () => null,
    renderBody: () => null,
    renderRight: () => null,
    disabled: false,
    accessibilityLabel: 'Encabezado',
    accessibilityHint: '',
    testID: 'Encabezado',
};

export default memo(HeaderHomeApp);
