// react-native-minerva/src/util/easing.js
import { Easing as EasingReactNative } from 'react-native';
import { Easing as EasingReactNativeReadnimated } from 'react-native-reanimated';

const Easing = {
    // Standard CSS easings

    linear: EasingReactNative.linear,
    ease: EasingReactNative.bezier(0.25, 0.1, 0.25, 1),
    ease_in: EasingReactNative.bezier(0.42, 0, 1, 1),
    ease_out: EasingReactNative.bezier(0, 0, 0.58, 1),
    ease_in_out: EasingReactNative.bezier(0.42, 0, 0.58, 1),

    // Penner Equations _ http://matthewlein.com/ceaser/ & http://easings.net

    ease_in_cubic: EasingReactNative.bezier(0.55, 0.055, 0.675, 0.19),
    ease_out_cubic: EasingReactNative.bezier(0.215, 0.61, 0.355, 1.0),
    ease_in_out_cubic: EasingReactNative.bezier(0.645, 0.045, 0.355, 1.0),

    ease_in_circ: EasingReactNative.bezier(0.6, 0.04, 0.98, 0.335),
    ease_out_circ: EasingReactNative.bezier(0.075, 0.82, 0.165, 1.0),
    ease_in_out_circ: EasingReactNative.bezier(0.785, 0.135, 0.15, 0.86),

    ease_in_expo: EasingReactNative.bezier(0.95, 0.05, 0.795, 0.035),
    ease_out_expo: EasingReactNative.bezier(0.19, 1.0, 0.22, 1.0),
    ease_in_out_expo: EasingReactNative.bezier(1.0, 0.0, 0.0, 1.0),

    ease_in_quad: EasingReactNative.bezier(0.55, 0.085, 0.68, 0.53),
    ease_out_quad: EasingReactNative.bezier(0.25, 0.46, 0.45, 0.94),
    ease_in_out_quad: EasingReactNative.bezier(0.455, 0.03, 0.515, 0.955),

    ease_in_quart: EasingReactNative.bezier(0.895, 0.03, 0.685, 0.22),
    ease_out_quart: EasingReactNative.bezier(0.165, 0.84, 0.44, 1.0),
    ease_in_out_quart: EasingReactNative.bezier(0.77, 0.0, 0.175, 1.0),

    ease_in_quint: EasingReactNative.bezier(0.755, 0.05, 0.855, 0.06),
    ease_out_quint: EasingReactNative.bezier(0.23, 1.0, 0.32, 1.0),
    ease_in_out_quint: EasingReactNative.bezier(0.86, 0.0, 0.07, 1.0),

    ease_in_sine: EasingReactNative.bezier(0.47, 0.0, 0.745, 0.715),
    ease_out_sine: EasingReactNative.bezier(0.39, 0.575, 0.565, 1.0),
    ease_in_out_sine: EasingReactNative.bezier(0.445, 0.05, 0.55, 0.95),

    ease_in_back: EasingReactNative.bezier(0.6, -0.28, 0.735, 0.045),
    ease_out_back: EasingReactNative.bezier(0.175, 0.885, 0.32, 1.275),
    ease_in_out_back: EasingReactNative.bezier(0.68, -0.55, 0.265, 1.55),
};

const EasingReanimated = {
    // Standard CSS easings

    linear: EasingReactNativeReadnimated.linear,
    ease: EasingReactNativeReadnimated.bezier(0.25, 0.1, 0.25, 1),
    ease_in: EasingReactNativeReadnimated.bezier(0.42, 0, 1, 1),
    ease_out: EasingReactNativeReadnimated.bezier(0, 0, 0.58, 1),
    ease_in_out: EasingReactNativeReadnimated.bezier(0.42, 0, 0.58, 1),

    // Penner Equations _ http://matthewlein.com/ceaser/ & http://easings.net

    ease_in_cubic: EasingReactNativeReadnimated.bezier(
        0.55,
        0.055,
        0.675,
        0.19,
    ),
    ease_out_cubic: EasingReactNativeReadnimated.bezier(
        0.215,
        0.61,
        0.355,
        1.0,
    ),
    ease_in_out_cubic: EasingReactNativeReadnimated.bezier(
        0.645,
        0.045,
        0.355,
        1.0,
    ),

    ease_in_circ: EasingReactNativeReadnimated.bezier(0.6, 0.04, 0.98, 0.335),
    ease_out_circ: EasingReactNativeReadnimated.bezier(0.075, 0.82, 0.165, 1.0),
    ease_in_out_circ: EasingReactNativeReadnimated.bezier(
        0.785,
        0.135,
        0.15,
        0.86,
    ),

    ease_in_expo: EasingReactNativeReadnimated.bezier(0.95, 0.05, 0.795, 0.035),
    ease_out_expo: EasingReactNativeReadnimated.bezier(0.19, 1.0, 0.22, 1.0),
    ease_in_out_expo: EasingReactNativeReadnimated.bezier(1.0, 0.0, 0.0, 1.0),

    ease_in_quad: EasingReactNativeReadnimated.bezier(0.55, 0.085, 0.68, 0.53),
    ease_out_quad: EasingReactNativeReadnimated.bezier(0.25, 0.46, 0.45, 0.94),
    ease_in_out_quad: EasingReactNativeReadnimated.bezier(
        0.455,
        0.03,
        0.515,
        0.955,
    ),

    ease_in_quart: EasingReactNativeReadnimated.bezier(
        0.895,
        0.03,
        0.685,
        0.22,
    ),
    ease_out_quart: EasingReactNativeReadnimated.bezier(0.165, 0.84, 0.44, 1.0),
    ease_in_out_quart: EasingReactNativeReadnimated.bezier(
        0.77,
        0.0,
        0.175,
        1.0,
    ),

    ease_in_quint: EasingReactNativeReadnimated.bezier(
        0.755,
        0.05,
        0.855,
        0.06,
    ),
    ease_out_quint: EasingReactNativeReadnimated.bezier(0.23, 1.0, 0.32, 1.0),
    ease_in_out_quint: EasingReactNativeReadnimated.bezier(
        0.86,
        0.0,
        0.07,
        1.0,
    ),

    ease_in_sine: EasingReactNativeReadnimated.bezier(0.47, 0.0, 0.745, 0.715),
    ease_out_sine: EasingReactNativeReadnimated.bezier(0.39, 0.575, 0.565, 1.0),
    ease_in_out_sine: EasingReactNativeReadnimated.bezier(
        0.445,
        0.05,
        0.55,
        0.95,
    ),

    ease_in_back: EasingReactNativeReadnimated.bezier(0.6, -0.28, 0.735, 0.045),
    ease_out_back: EasingReactNativeReadnimated.bezier(
        0.175,
        0.885,
        0.32,
        1.275,
    ),
    ease_in_out_back: EasingReactNativeReadnimated.bezier(
        0.68,
        -0.55,
        0.265,
        1.55,
    ),
};

export { EasingReanimated, Easing };
