// react-native-minerva/src/util/getStatusBarHeight.js
import { Dimensions, Platform, StatusBar } from 'react-native';

const { height } = Dimensions.get('window');

const defaultiPhoneStatusBarHeight = 20;

export const iPhoneWithNotch = [
    {
        width: 375,
        height: 812,
        statusBarHeight: 44,
    },
    {
        width: 390,
        height: 844,
        statusBarHeight: 47,
    },
    {
        width: 393,
        height: 852,
        statusBarHeight: 59,
    },
    {
        width: 414,
        height: 896,
        statusBarHeight: 44,
    },
    {
        width: 428,
        height: 926,
        statusBarHeight: 47,
    },
    {
        width: 430,
        height: 932,
        statusBarHeight: 59,
    },
];

const isiPhoneWithNotch = iPhoneWithNotch.find((key) => key.height === height);

const iphoneStatusBarHeight = isiPhoneWithNotch
    ? isiPhoneWithNotch.statusBarHeight
    : defaultiPhoneStatusBarHeight;

/**
 * Función para obtener altura del statusbar.
 *
 * @param {boolean} skipAndroid - Boleano que retorna 0 altura para android.
 * @returns {object} {} - Retorna objecto para configurar altura por plataforma.
 */
export function getStatusBarHeight(skipAndroid) {
    return Platform.select({
        ios: iphoneStatusBarHeight,
        android: skipAndroid ? 0 : StatusBar.currentHeight,
        default: 0,
    });
}

export default getStatusBarHeight;
