// react-native-minerva/src/components/chip/components/ChipText.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Text, HelperStyle } from '../../../index';
import buildStyle from '../style';
import { chipPropTypes, chipDefaultProps } from '../type';

const ChipText = ({
    avatar,
    badge,
    bigger,
    children,
    icon,
    removable,
    selected,
    textProps,
}) => {
    const styles = buildStyle(HelperStyle);
    const textStyle = [styles.defaultStyleText, selected && styles.activeText];
    textStyle.push(icon && styles.textIcon);
    textStyle.push(avatar && styles.textAvatar);
    textStyle.push(badge && styles.textBadge);
    textStyle.push(bigger && styles.textBigger);
    textStyle.push(removable && styles.textRemovable);
    return (
        <Text small style={textStyle} {...textProps}>
            {children}
        </Text>
    );
};

ChipText.propTypes = {
    avatar: chipPropTypes.avatar,
    badge: chipPropTypes.badge,
    bigger: chipPropTypes.bigger,
    children: PropTypes.string.isRequired,
    icon: chipPropTypes.icon,
    removable: chipPropTypes.removable,
    selected: chipPropTypes.selected,
    textProps: chipPropTypes.textProps,
};
ChipText.defaultProps = {
    avatar: chipDefaultProps.avatar,
    badge: chipDefaultProps.badge,
    bigger: chipDefaultProps.bigger,
    icon: chipDefaultProps.icon,
    removable: chipDefaultProps.removable,
    selected: chipDefaultProps.selected,
    textProps: chipDefaultProps.textProps,
};

export default memo(ChipText);
