// react-native-minerva/src/components/linearGradient/LinearGradient.js
import React, { memo } from 'react';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import LinearGradient from 'react-native-linear-gradient';
import { HelperStyle } from '../../index';

/**
 * Componente para mostrar gradiente. Extiende de react-native-linear-gradient.
 *
 * @component
 */
const LinearGradientBCH = ({
    children,
    colors,
    end,
    locations,
    start,
    style,
}) => {
    const colorsGradient = isEmpty(colors)
        ? [HelperStyle.brand_light.color, HelperStyle.brand_dark.color]
        : colors;
    return (
        <LinearGradient
            colors={colorsGradient}
            end={end}
            locations={locations}
            start={start}
            style={style}
        >
            {children}
        </LinearGradient>
    );
};

LinearGradientBCH.propTypes = {
    /**
     * Renderiza componente.
     */
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.bool]),
    /**
     * Arreglo de colores para el gradiente.
     */
    colors: PropTypes.arrayOf(PropTypes.string),
    /**
     * Estilos.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Arreglo de números que define la ubicación de cada color del degradado
     * Ejemplo: [0.1, 0.75, 1] significa que el primer color ocupará 0% - 10% el segundo color ocupará 10% - 75% y el tercer color ocupará 75% - 100%.
     */
    locations: PropTypes.arrayOf(PropTypes.number),
    /**
     * Objeto opcional del siguiente tipo: { x: número, y: número }. Con las coordenadas de la posición en la que comienza el degradado.
     */
    start: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
    }),
    /**
     * Objeto opcional del siguiente tipo: { x: número, y: número }. Con las coordenadas de la posición en la que termina el degradado.
     */
    end: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
    }),
};
LinearGradientBCH.defaultProps = {
    children: false,
    colors: [],
    style: {},
    locations: [0.0, 0.99],
    start: { x: 0, y: 0 },
    end: { x: 1, y: 1 },
};

export default memo(LinearGradientBCH);
