// bch-mobile-tooltip/src/Onboarding.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import {
    Icon,
    HelperStyle,
    Button,
    Grid,
    Row,
    Col,
    Text,
} from 'react-native-minerva';
import labels from './labelAutomatizacion';

const propTypes = {
    title: PropTypes.string,
    content: PropTypes.string,
    status: PropTypes.string,
    onPressClose: PropTypes.func.isRequired,
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.func,
    ]),
    style: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
    ]),
};

const defaultProps = {
    title: '',
    content: '',
    status: '',
    style: {},
    children: false,
};

const Onboarding = ({
    title,
    content,
    status,
    children,
    onPressClose,
    style,
}) => {
    return (
        <Grid style={[HelperStyle.minw_100, HelperStyle.as_c, HelperStyle.ai_c, style]}>
            <Row>
                { title && (
                    <Col size={9} style={HelperStyle.ac_c}>
                        <Text
                            bold
                            style={[HelperStyle.mb_2, HelperStyle.gray]}
                            accessible
                            accessibilityLabel={title}
                            testID={title}
                            accessibilityRole="text"
                        >
                            {title}
                        </Text>
                    </Col>
                )}
                <Col size={1} style={HelperStyle.jc_c}>
                    <Button
                        link
                        secondary
                        smaller
                        style={[HelperStyle.as_fe, HelperStyle.pb_1]}
                        onPress={onPressClose}
                        accessibilityLabel={labels.CERRAR}
                        accessibilityHint={labels.CERRAR}
                        testID={labels.CERRAR}
                        accessibilityRole="button"
                    >
                        <Icon
                            style={[HelperStyle.gray, HelperStyle.fs_2]}
                            name="icon-bch-close-strong"
                        />
                    </Button>
                </Col>
            </Row>
            { content && (
                <Row>
                    <Col size={10}>                       
                        <Text
                            smaller
                            style={HelperStyle.mb_2}
                            accessibilityLabel={content}
                            testID={content}
                            accessibilityRole="text"
                        >
                            {content}
                        </Text>                        
                    </Col>
                </Row>
            )}
            <Row>
                <Col size={2} style={HelperStyle.jc_c}>
                    <Text
                        small
                        style={HelperStyle.mb_0}
                        accessibilityLabel={status}
                        testID={status}
                        accessibilityRole="text"
                    >
                        {status}
                    </Text>
                </Col>
                {/* botones de navegacion */}
                <Col size={8} style={[HelperStyle.flx_row, HelperStyle.jc_fe]}>
                    {children}                     
                </Col>
            </Row>
        </Grid>
    );
};

Onboarding.propTypes = propTypes;
Onboarding.defaultProps = defaultProps;

export default memo(Onboarding);
