// react-native-minerva/src/components/pagination/type.js
import PropTypes from 'prop-types';

const dotsPropTypes = {
    /**
     * Boleano para definir estado activo.
     */
    active: PropTypes.bool,
    /**
     * Color del estado activo.
     */
    activeColor: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Color del estado inactivo.
     */
    inactiveColor: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    /**
     * Estilos.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Identificador de pruebas para el dot.
     */
    testID: PropTypes.string,
    /**
     * Identificador de accesibilidad para el dot.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad para el dot.
     */
    accessibilityHint: PropTypes.string,
    /**
     * Boleano para habilitar accesibilidad para el dot.
     */
    accessible: PropTypes.bool,
};

const dotsDefaultProps = {
    accessibilityLabel: '',
    accessibilityHint: '',
    accessible: false,
    active: false,
    activeColor: false,
    inactiveColor: false,
    style: {},
    testID: '',
};

const paginationPropTypes = {
    /**
     * Índice del activo.
     */
    activeDotIndex: PropTypes.number,
    /**
     * Numero de paginas.
     */
    dotsLength: PropTypes.number,
    /**
     * Tema de la paginación light o dark.
     */
    paginationStyle: PropTypes.oneOf(['light', 'dark']),
    /**
     * Estilos para el contendor de la paginación.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Identificador de pruebas para el contenedor del paginador.
     */
    testID: PropTypes.string,
};

const paginationDefaultProps = {
    testID: '',
    dotsLength: 5,
    activeDotIndex: 0,
    paginationStyle: 'dark',
    style: {},
};

export {
    dotsPropTypes,
    dotsDefaultProps,
    paginationPropTypes,
    paginationDefaultProps,
};
