// react-native-minerva/src/components/collapsingHeader/style.js
import { Platform, StyleSheet } from 'react-native';
import {
    getHeaderHeight,
    getHeaderStatusBarHeight,
    dimensions,
} from '../../index';

const isAndroid = Platform.OS === 'android';

const Style = ({ minHeight, heightCircle }) =>
    StyleSheet.create({
        overlapping: {
            marginTop: -heightCircle,
        },
        bounceStyle: {
            minHeight,
        },
        translateStyle: {
            paddingBottom: heightCircle,
        },
        headerFixedForeground: {
            position: 'relative',
            height: getHeaderHeight,
            zIndex: 1,
        },
        imageBackgroundHeaderSmall: {
            width: dimensions.fullWidth,
            height: isAndroid ? getHeaderStatusBarHeight : getHeaderHeight,
            position: 'absolute',
            top: 0,
            zIndex: 0,
        },
        imageBackgroundHeaderDefault: {
            width: '100%',
            height: '100%',
            position: 'absolute',
            top: 0,
            zIndex: 0,
            transform: [{ scaleX: 0.5 }, { translateY: 0.5 }],
        },
        imageStyle: {
            width: '100%',
            height: '100%',
            resizeMode: 'cover',
        },
    });

export default Style;
