// react-native-minerva/src/components/carousel/components/ButtonLastItem.js
import React, { memo } from 'react';
import { HelperStyle, View, dimensions, Button, Text } from '../../../index';
import CAROUSEL_STR from '../CarouselStr';
import { buttonLastItemPropTypes, buttonLastItemDefaultProps } from '../types';

const calcularAncho = (percentage) => {
    const value = (percentage * dimensions.fullWidth) / 100;
    return Math.round(value);
};

const ButtonLastItem = (props) => {
    const {
        activeSlideAlignment,
        title,
        style,
        testID,
        accessibilityLabel,
        accessibilityHint,
    } = props;
    const slideWidth = calcularAncho(75);
    const itemHorizontalMargin = calcularAncho(2);
    const sliderWidth = dimensions.fullWidth;
    const paddingCarousel = sliderWidth - slideWidth - itemHorizontalMargin * 2;
    const styleContentButton =
        activeSlideAlignment === 'start'
            ? { marginLeft: paddingCarousel - 16 }
            : null;
    return (
        <View
            style={[
                HelperStyle.jc_c,
                HelperStyle.ai_c,
                HelperStyle.flx_i,
                styleContentButton,
            ]}
        >
            <Button
                small
                primary
                iconName={CAROUSEL_STR.buttonLastItem.iconName}
                {...props}
                testID={testID}
                accessibilityLabel={accessibilityLabel}
                accessibilityHint={accessibilityHint}
                style={[HelperStyle.shadow, HelperStyle.as_c, style]}
            >
                <Text>{title}</Text>
            </Button>
        </View>
    );
};

ButtonLastItem.propTypes = {
    activeSlideAlignment: buttonLastItemPropTypes.activeSlideAlignment,
    title: buttonLastItemPropTypes.title,
    style: buttonLastItemPropTypes.style,
    testID: buttonLastItemPropTypes.testID,
    accessibilityLabel: buttonLastItemPropTypes.accessibilityLabel,
    accessibilityHint: buttonLastItemPropTypes.accessibilityHint,
};
ButtonLastItem.defaultProps = {
    activeSlideAlignment: buttonLastItemDefaultProps.activeSlideAlignment,
    title: buttonLastItemDefaultProps.title,
    style: buttonLastItemDefaultProps.style,
    testID: buttonLastItemDefaultProps.testID,
    accessibilityLabel: buttonLastItemDefaultProps.accessibilityLabel,
    accessibilityHint: buttonLastItemDefaultProps.accessibilityHint,
};

export default memo(ButtonLastItem);
