// react-native-minerva/src/components/checkBox/type.js
import PropTypes from 'prop-types';
import { DEFAULT_TYPES, TYPES } from '../../constants/propTypes';

const propTypes = {
    /**
     * Controla el evento touch.
     */
    pointerEvents: PropTypes.string,
    /**
     * Etiqueta del componente.
     */
    label: PropTypes.string,
    /**
     * Identificador de accesibilidad para label.
     */
    labelAccessibilityLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad que ayuda a los usuarios a comprender qué sucederá cuando realicen una acción.
     */
    labelAccessibilityHint: PropTypes.string,
    /**
     * Identificador de pruebas para label.
     */
    labelTestID: PropTypes.string,
    /**
     * Estado deshabilitado.
     */
    disabled: PropTypes.bool,
    /**
     * Estilo en linea.
     */
    inline: PropTypes.bool,
    /**
     * Estilo para checkbox.
     */
    style: TYPES.style,
    /**
     * Estilo para View container del checkbox.
     */
    styleContainer: TYPES.style,
    /**
     * Componente a renderizar en el area del label.
     */
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.string,
        PropTypes.func,
    ]),
    /**
     * El estado del checkbox (selecciona / deselecciona).
     */
    checked: PropTypes.bool,
    /**
     * Función que se llama cuando el usuario selecciona / deselecciona el checkbox.
     */
    onPress: PropTypes.func,
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Identificador para accesibilidad.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Accesibilidad: ayuda a los usuarios a comprender qué sucederá cuando realicen una acción en el elemento.
     */
    accessibilityHint: PropTypes.string,
    /**
     * Comunica el propósito de un componente al usuario.
     */
    accessibilityRole: PropTypes.string,
    /**
     * Para customizar el icono que se muestra cuando esta seleccionado.
     * Se debe usar la libreria de iconos Ionicons https://ionic.io/ionicons.
     */
    iconName: PropTypes.string,
};

const defaultProps = {
    accessibilityHint: '',
    accessibilityLabel: '',
    accessibilityRole: 'checkbox',
    checked: false,
    children: null,
    disabled: false,
    inline: false,
    label: '',
    labelAccessibilityLabel: '',
    labelAccessibilityHint: '',
    labelTestID: '',
    onPress: DEFAULT_TYPES.func,
    pointerEvents: 'auto',
    style: DEFAULT_TYPES.style,
    styleContainer: DEFAULT_TYPES.style,
    testID: '',
    iconName: '',
};

export { propTypes, defaultProps };
