// react-native-minerva/src/components/subTabs/type.js
import PropTypes from 'prop-types';

const subTabsHeaderPropTypes = {
    /**
     * Función para cambiar el estado de itemActive.
     */
    selected: PropTypes.func,
    /**
     * Índice del activo.
     */
    itemActive: PropTypes.number,
    /**
     * Arreglo de objectos para configurar los subTabs.
     */
    buttonsHeader: PropTypes.arrayOf(
        PropTypes.shape({
            /**
             * Título del subtab.
             */
            title: PropTypes.string,
            /**
             * Identificador para pruebas.
             */
            testID: PropTypes.string,
            /**
             * Identificador para accesibilidad.
             */
            accessibilityLabel: PropTypes.string,
            /**
             * Sugerencia de accesibilidad.
             */
            accessibilityHint: PropTypes.string,
        }),
    ),
    /**
     * Estilos del contenedor del flatlist.
     */
    contentContainerStyle: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.object,
    ]),
    /**
     * Boleano para definir botón subtab tamaño small.
     */
    small: PropTypes.bool,
    /**
     * Boleano para agregar sombra al subtab.
     */
    shadow: PropTypes.bool,
    /**
     * Boleano para agregar scroll horizontal.
     */
    scroll: PropTypes.bool,
    /**
     * Estilos para <Segment />.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Comunica el propósito de un componente al usuario.
     */
    accessibilityRole: PropTypes.string,
};

const subTabsHeaderDefaultProps = {
    selected: () => null,
    itemActive: 1,
    buttonsHeader: [],
    contentContainerStyle: {},
    small: false,
    shadow: true,
    scroll: false,
    style: {},
    testID: '',
    accessibilityRole: 'tab',
};

const subTabsPropTypes = {
    renderContentSegment: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.element,
        PropTypes.node,
    ]),
};

const subTabsDefaultProps = {
    renderContentSegment: null,
};

export {
    subTabsHeaderPropTypes,
    subTabsHeaderDefaultProps,
    subTabsPropTypes,
    subTabsDefaultProps,
};
