// bch-mobile-menu/src/types.js
import PropTypes from 'prop-types';

const menuOptions = PropTypes.arrayOf(
    PropTypes.shape({
        showCard: PropTypes.bool,
        newOptionIcon: PropTypes.bool,
        icon: PropTypes.string,
        title: PropTypes.string,
        action: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.func]),
        collapsed: PropTypes.bool,
        options: PropTypes.arrayOf(
            PropTypes.shape({
                text: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
                action: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.func]),
                newOption: PropTypes.bool,
            }),
        ),
    }),
);

const actionCerrarMenu = PropTypes.func;

const user = PropTypes.shape({
    name: PropTypes.string,
    lastName: PropTypes.string,
    onPressEditLink: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
    onPressAvatar: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
    loadingAvatar: PropTypes.bool,
    image: PropTypes.string,
});

const hiddenAvatar = PropTypes.bool;
const disabledHeaderTouch = PropTypes.bool;

const renderMenuFooter = PropTypes.oneOfType([PropTypes.func, PropTypes.bool]);

export const menuPropTypes = {
    route: PropTypes.shape({
        params: PropTypes.shape({
            menuOptions,
            actionCerrarMenu,
            user,
            renderMenuFooter,
            hiddenAvatar,
        }),
    }),
    navigation: PropTypes.shape({
        navigate: PropTypes.func,
        closeDrawer: PropTypes.func,
    }),
    menuOptions,
    actionCerrarMenu,
    renderMenuFooter,
    user,
    hiddenAvatar,
    disabledHeaderTouch,
};

export const menuDefaultProps = {
    menuOptions: [],
    actionCerrarMenu: () => null,
    user: {
        name: '',
        lastName: '',
        onPressEditLink: undefined,
        onPressAvatar: undefined,
        loadingAvatar: false,
        image: '',
    },
    hiddenAvatar: false,
    renderMenuFooter: false,
    disabledHeaderTouch: false,
    navigation: null,
};
