// react-native-minerva/src/components/passwordChecklist/PasswordChecklist.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { MapList } from '../../index';
import PasswordCheckitem from './PasswordCheckitem';

/**
 * Componente listado de políticas de seguridad para PIN.
 *
 * @component
 * @example
 * const data = [ {value: 'Condición 1', check: true} ];
 * return (
 *   <PasswordChecklist data={data}/>
 * );
 */
const PasswordChecklist = ({ data }) => {
    const renderCheckItem = ({
        item: { value, check, testID, accessibilityLabel, accessibilityHint },
    }) => (
        <PasswordCheckitem
            value={value}
            check={check}
            testID={testID}
            accessibilityLabel={accessibilityLabel}
            accessibilityHint={accessibilityHint}
        />
    );
    return <MapList data={data} renderItem={renderCheckItem} wrapper={false} />;
};

PasswordChecklist.propTypes = {
    /**
     * Arreglo de políticas de seguridad para PIN.
     */
    data: PropTypes.arrayOf(
        PropTypes.shape({
            /**
             * Texto políticas de seguridad para PIN.
             */
            value: PropTypes.string.isRequired,
            /**
             * Estado políticas de Seguridad para PIN.
             */
            check: PropTypes.bool.isRequired,
            /**
             * Identificador para pruebas.
             */
            testID: PropTypes.string,
            /**
             * Identificador para accesibilidad.
             */
            accessibilityLabel: PropTypes.string,
            /**
             * Sugerencia de accesibilidad que ayuda a los usuarios a comprender qué sucederá cuando realicen una acción.
             */
            accessibilityHint: PropTypes.string,
        }),
    ).isRequired,
};

export default memo(PasswordChecklist);
