// @bch/eslint-config/index.js
module.exports = {
    parser: '@babel/eslint-parser',
    env: {
        es2021: true,
    },
    extends: [
        'airbnb',
        'eslint:recommended',
        'plugin:react/recommended',
        'plugin:sonarjs/recommended',
        'plugin:prettier/recommended',
    ],
    parserOptions: {
        ecmaFeatures: {
            jsx: true,
        },
        ecmaVersion: 13,
        sourceType: 'module',
    },
    plugins: ['react', 'react-native', 'react-hooks', 'sonarjs', 'prettier'],
    rules: {
        // Reglas para eslint
        'arrow-body-style': 'warn',
        'linebreak-style': 'off',
        'max-len': ['error', { code: 200 }],
        'no-underscore-dangle': [
            'error',
            {
                allow: ['_root'],
            },
        ],
        'no-param-reassign': 0,
        'no-shadow': 'off',
        'operator-linebreak': 'off',
        'object-curly-newline': 'off',
        // Reglas para eslint-config-prettier
        'prettier/prettier': 'error',
        // Reglas para eslint-plugin-react
        'react/jsx-closing-bracket-location': [1, 'line-aligned'],
        'react/no-multi-comp': [
            'error',
            {
                ignoreStateless: true,
            },
        ],
        'react/jsx-filename-extension': [
            1,
            {
                extensions: ['.js'],
            },
        ],
        'react/function-component-definition': [
            // Agrega arrow-function a regla de eslint-config-airbnb https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/rules/react.js#L527
            2,
            {
                namedComponents: ['function-declaration', 'function-expression', 'arrow-function'],
                unnamedComponents: ['function-expression', 'arrow-function'],
            },
        ],
        // Reglas para eslint-plugin-react-hooks
        'react-hooks/rules-of-hooks': 'error',
        'react-hooks/exhaustive-deps': 'off',
        // Reglas para eslint-plugin-import
        'import/no-named-as-default': 'off',
        'import/no-extraneous-dependencies': 0,
        // Reglas para eslint-plugin-react-native
        'react-native/no-unused-styles': 2,
        'react-native/split-platform-components': 2,
        'react-native/no-inline-styles': 2,
        'react-native/no-color-literals': 2,
        'react-native/no-raw-text': [
            2,
            {
                skip: [
                    'H1',
                    'H2',
                    'H3',
                    'H4',
                    'H5',
                    'Animated.Text',
                    'InputHelperText',
                ],
            },
        ],
        'react-native/no-single-element-style-arrays': 2,
    },
    globals: {
        fetch: false,
        jest: false,
        test: false,
        expect: false,
        describe: false,
        beforeEach: false,
        afterEach: false,
        window: false,
        __DEV__: false,
    },
    settings: {
        'import/resolver': {
            node: {
                extensions: ['.js'],
                paths: ['src'],
            },
        },
    },
    overrides: [
        {
            files: ['**/*.test.js'],
            env: {
                jest: true,
            },
        },
    ],
};
