// bch-mobile-skeleton/index.js
/*
TODO:
    eliminar skeleton bch-mobile-common-component
    eliminar de bch-mobile-movimientos-cuenta
*/
import SkeletonBullets from './src/components/SkeletonBullets';
import SkeletonCardProducto from './src/components/SkeletonCardProducto';
import SkeletonMultiLinea from './src/components/SkeletonMultiLinea';
import SkeletonPillsBotonAgregar from './src/components/SkeletonPillsBotonAgregar';
import SkeletonSaldoConsolidado from './src/components/SkeletonSaldoConsolidado';
import SkeletonUltimosMovimientos from './src/components/SkeletonUltimosMovimientos';
import SkeletonRubrosDestacados from './src/components/SkeletonRubrosDestacados';
import SkeletonRubrosNoDestacados from './src/components/SkeletonRubrosNoDestacados';
import SkeletonListaEmpresas from './src/components/SkeletonListaEmpresas';
import SkeletonMovimientos from './src/components/SkeletonMovimientos';
import SkeletonCardsResumen from './src/components/SkeletonCardsResumen';
import SkeletonTarjetaFan from './src/components/SkeletonTarjetaFan';
import SkeletonDispositivosRegistrados from './src/components/SkeletonDispositivosRegistrados';
import SkeletonHeaderHistoricoCuentas from './src/components/SkeletonHeaderHistoricoCuentas';
import SkeletonContentHistoricoCuentas from './src/components/SkeletonContentHistoricoCuentas';
import SkeletonHeaderDetalleCuenta from './src/components/SkeletonHeaderDetalleCuenta';
import SkeletonContentDetalleCuenta from './src/components/SkeletonContentDetalleCuenta';
import SkeletonResumenCuentaPrincipal from './src/components/SkeletonResumenCuentaPrincipal';
import SkeletonLineaCuentaPrincipal from './src/components/SkeletonLineaCuentaPrincipal';
import SkeletonMisCuentas from './src/views/SkeletonMisCuentas';
import SkeletonAgregarCuentaRubros from './src/views/SkeletonAgregarCuentaRubros';
import SkeletonCompraDolaresPorCaja from './src/views/SkeletonCompraDolaresPorCaja';

export {
    SkeletonBullets,
    SkeletonCardProducto,
    SkeletonMultiLinea,
    SkeletonPillsBotonAgregar,
    SkeletonSaldoConsolidado,
    SkeletonUltimosMovimientos,
    SkeletonRubrosDestacados,
    SkeletonRubrosNoDestacados,
    SkeletonListaEmpresas,
    SkeletonMovimientos,
    SkeletonCardsResumen,
    SkeletonTarjetaFan,
    SkeletonDispositivosRegistrados,
    SkeletonHeaderHistoricoCuentas,
    SkeletonContentHistoricoCuentas,
    SkeletonHeaderDetalleCuenta,
    SkeletonContentDetalleCuenta,
    SkeletonLineaCuentaPrincipal,
    SkeletonResumenCuentaPrincipal,
    SkeletonMisCuentas,
    SkeletonAgregarCuentaRubros,
    SkeletonCompraDolaresPorCaja,
};
