// react-native-minerva/src/components/squareButtonsGroup/components/SquareButton.js
import React, { memo } from 'react';
import { Pressable } from 'react-native';
import PropTypes from 'prop-types';
import {
    Body,
    Card,
    CardItem,
    HelperStyle,
    Icon,
    Text,
    View,
} from '../../../index';
import buildStyle from '../style';

/**
 * Componente usado por @SquareButtonsGroup y no exportado(no se puede usar directamente).
 *
 * @component
 */
const SquareButton = ({
    insideText,
    centered,
    style,
    iconName,
    iconStyle,
    text,
    textStyle,
    onPress,
    selected,
    testID,
    accessibilityLabel,
    accessibilityHint,
}) => {
    const Style = buildStyle(HelperStyle, centered, insideText, selected);

    return (
        <View style={[Style.item, style]}>
            <Card bordered style={Style.itemStyle}>
                <CardItem
                    button
                    onPress={onPress}
                    testID={testID}
                    accessibilityLabel={accessibilityLabel}
                    accessibilityHint={accessibilityHint}
                    accessibilityRole="button"
                >
                    <Body style={Style.bodyCard}>
                        <Icon
                            name={iconName}
                            style={[
                                Style.iconStyle,
                                iconStyle,
                                selected && HelperStyle.primary,
                            ]}
                        />
                        {insideText && (
                            <Text smaller style={[Style.textInside, textStyle]}>
                                {text}
                            </Text>
                        )}
                    </Body>
                </CardItem>
            </Card>
            {!insideText && (
                <Pressable
                    accessibilityRole="link"
                    onPress={onPress}
                    style={HelperStyle.as_c}
                >
                    <Text smaller style={[Style.textOutside, textStyle]}>
                        {text}
                    </Text>
                </Pressable>
            )}
        </View>
    );
};

SquareButton.propTypes = {
    /**
     * Nombre del icono segun el asignado en bch-assets.
     */
    iconName: PropTypes.string.isRequired,
    /**
     * (Opcional)Estilo personalizado para el icono.
     */
    iconStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Texto a mostrar.
     */
    text: PropTypes.string.isRequired,
    /**
     * (Opcional)Estilo personalizado para el texto.
     */
    textStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Funcion que define la logica cuando la card sea pulsada.
     */
    onPress: PropTypes.func.isRequired,
    /**
     * (Opcional) Estilo personalizado.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    /**
     * Flag que define si el texto se muestra dentro o debajo de la card.
     */
    insideText: PropTypes.bool,
    /**
     * Flag que define si es el ultimo elemento y es el unico de su fila.
     */
    centered: PropTypes.bool,
    /**
     * Flag que define si la card tiene estado selected/unselected.
     */
    selected: PropTypes.bool,
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Identificador para accesibilidad.
     */
    accessibilityLabel: PropTypes.string.isRequired,
    /**
     * Sugerencia de accesibilidad.
     */
    accessibilityHint: PropTypes.string.isRequired,
};
SquareButton.defaultProps = {
    iconStyle: {},
    textStyle: {},
    style: {},
    insideText: false,
    centered: false,
    selected: false,
    testID: '',
};

export default memo(SquareButton);
