// react-native-minerva/src/components/paragraph/types.js
import PropTypes from 'prop-types';
import { DEFAULT_TYPES, TYPES } from '../../constants/propTypes';

const propTypes = {
    /**
     * Arreglo de textos o objecto con propiedades para configurar textos y sus proppiedades.
     */
    items: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
            /**
             * Texto.
             */
            text: PropTypes.string,
            /**
             * Identificador de accesibilidad.
             */
            accessibilityLabel: PropTypes.string,
            /**
             * Sugerencia de accesibilidad.
             */
            accessibilityHint: PropTypes.string,
            /**
             * Si esta definido indica el numero de lineas que puede tener el texto antes de truncarlo.
             */
            numberOfLines: PropTypes.number,
            /**
             * Cuando numberOfLines esta definido esta propiedad define c칩mo se truncar치 el texto.
             */
            ellipsizeMode: PropTypes.oneOf(['head', 'middle', 'tail', 'clip']),
            /**
             * Identificador de pruebas.
             */
            testID: PropTypes.string,
            /**
             * Estilos.
             */
            style: TYPES.style,
        }),
        PropTypes.arrayOf(
            PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number,
                PropTypes.shape({
                    /**
                     * Texto.
                     */
                    text: PropTypes.string,
                    /**
                     * Identificador de accesibilidad.
                     */
                    accessibilityLabel: PropTypes.string,
                    /**
                     * Sugerencia de accesibilidad.
                     */
                    accessibilityHint: PropTypes.string,
                    /**
                     * Si esta definido indica el numero de lineas que puede tener el texto antes de truncarlo.
                     */
                    numberOfLines: PropTypes.number,
                    /**
                     * Cuando numberOfLines esta definido esta propiedad define c칩mo se truncar치 el texto.
                     */
                    ellipsizeMode: PropTypes.oneOf([
                        'head',
                        'middle',
                        'tail',
                        'clip',
                    ]),
                    /**
                     * Identificador de pruebas.
                     */
                    testID: PropTypes.string,
                    /**
                     * Estilos.
                     */
                    style: TYPES.style,
                }),
            ]),
        ),
    ]),
    /**
     * Boleano para habilitar y deshabilitar.
     */
    disabled: PropTypes.bool,
    /**
     * Boleano para definir si envuelve los item en un View o no.
     */
    wrapper: PropTypes.bool,
    /**
     * Estilos que se aplica a todos los <Text /> dentro del arreglo de items.
     */
    style: TYPES.style,
    /**
     * Estilos del wrapper.
     */
    wrapperStyle: TYPES.style,
};

const defaultProps = {
    items: '',
    disabled: false,
    style: DEFAULT_TYPES.style,
    wrapper: true,
    wrapperStyle: DEFAULT_TYPES.style,
};

export { propTypes, defaultProps };
