// react-native-minerva/src/components/switch/type.js
import PropTypes from 'prop-types';

const SwitchpropTypes = {
    /**
     * Label a mostrar en el lado izquierdo del switch.
     */
    label: PropTypes.string,
    /**
     * Boleano para cambiar la direccion del label al lado derecho.
     */
    labelRight: PropTypes.bool,
    /**
     * Renderiza componente personalizado en el area del label.
     */
    renderLabel: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
    /**
     * Boleano para activar o desactivar el switch.
     */
    disabled: PropTypes.bool,
    /**
     * Boleano para indicar si esta activo o no el switch.
     */
    value: PropTypes.bool,
    /**
     * Boleano para indicar si esta cargando o no el switch.
     */
    loading: PropTypes.bool,
    /**
     * Estilos personalizados.
     */
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

const SwitchdefaultProps = {
    label: '',
    labelRight: false,
    renderLabel: false,
    disabled: false,
    value: false,
    loading: false,
    style: null,
};

export { SwitchpropTypes, SwitchdefaultProps };
