// react-native-minerva/src/components/cardMenu/Components/ItemLeft.js
import React, { memo } from 'react';
import { Image } from 'react-native';
import { HelperStyle, Icon, Validate } from '../../../index';
import { itemLeftTypes, itemLeftDefaultTypes } from '../types';
import Styles from './style';

const ItemLeft = ({
    iconStyle,
    iconName,
    iconLabel,
    imageSource,
    iconAccessibilityHint,
    iconTestID,
}) => {
    if (iconName) {
        return (
            <Icon
                name={iconName}
                accessibilityLabel={iconLabel}
                accessibilityHint={iconAccessibilityHint}
                testID={Validate.safeLabel(iconTestID, iconLabel)}
                style={[HelperStyle.pa_2, iconStyle]}
            />
        );
    }
    if (imageSource) {
        return <Image source={imageSource} style={Styles.image} />;
    }
    return null;
};

ItemLeft.propTypes = {
    iconName: itemLeftTypes.iconName,
    iconLabel: itemLeftTypes.iconLabel,
    iconAccessibilityHint: itemLeftTypes.iconAccessibilityHint,
    iconTestID: itemLeftTypes.iconTestID,
    iconStyle: itemLeftTypes.iconStyle,
    imageSource: itemLeftTypes.imageSource,
};
ItemLeft.defaultProps = {
    iconLabel: itemLeftDefaultTypes.iconLabel,
    iconAccessibilityHint: itemLeftDefaultTypes.iconAccessibilityHint,
    iconTestID: itemLeftDefaultTypes.iconTestID,
    iconName: itemLeftDefaultTypes.iconName,
    iconStyle: itemLeftDefaultTypes.iconStyle,
    imageSource: itemLeftDefaultTypes.imageSource,
};

export default memo(ItemLeft);
