// bch-mobile-menu/src/navigation/MenuNavigator.js
import React from 'react';
import { createStackNavigator } from '@react-navigation/stack';
import Menu from '../containers/Menu';
import { menuPropTypes, menuDefaultProps } from '../types';

const Stack = createStackNavigator();

const propTypes = {
    route: menuPropTypes.route,
};

const defaultProps = {
    route: {
        params: {
            menuOptions: menuDefaultProps.menuOptions,
            actionCerrarMenu: menuDefaultProps.actionCerrarMenu,
            user: menuDefaultProps.user,
            renderMenuFooter: menuDefaultProps.renderMenuFooter,
            disabledHeaderTouch: menuDefaultProps.disabledHeaderTouch,
        },
    },
};

const MenuNavigator = (props) => {
    const { route } = props;
    const { menuOptions, actionCerrarMenu, user, renderMenuFooter, hiddenAvatar, disabledHeaderTouch } = route.params;

    return (
        <Stack.Navigator initialRouteName="Menu" screenOptions={{ gestureEnabled: false }}>
            <Stack.Screen
                initialParams={{
                    menuOptions,
                    actionCerrarMenu,
                    user,
                    renderMenuFooter,
                    hiddenAvatar,
                    disabledHeaderTouch,
                }}
                name="Menu"
                component={Menu}
                options={{ headerShown: false }}
            />
        </Stack.Navigator>
    );
};

MenuNavigator.propTypes = propTypes;
MenuNavigator.defaultProps = defaultProps;

export default MenuNavigator;
