// react-native-minerva/src/components/pressableOpacity/type.js
import PropTypes from 'prop-types';
import { DEFAULT_TYPES, TYPES } from '../../constants/propTypes';

const propTypes = {
    /**
     * Estilos.
     */
    style: TYPES.style,
    /**
     * Componente a renderizar dentro de la caja.
     */
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
    /**
     * Duración (en milisegundos) desde el onPressIn antes de que se llame a onLongPress.
     */
    delayLongPress: PropTypes.number,
    /**
     * Boleano para habilitar y deshabilitar la caja.
     */
    disabled: PropTypes.bool,
    /**
     * Establece una distancia adicional fuera del elemento en la que se puede detectar el evento press.
     */
    hitSlop: PropTypes.shape({
        bottom: PropTypes.number,
        left: PropTypes.number,
        right: PropTypes.number,
        top: PropTypes.number,
    }),
    /**
     * Función que se llama al tocar por mas de 500 milisegundos.
     */
    onLongPress: PropTypes.func,
    /**
     * Función que se llama al tocar el componente.
     */
    onPress: PropTypes.func,
    /**
     * Función que se llama inmediatamente cuando se activa un toque, antes de onPressOut y onPress.
     */
    onPressIn: PropTypes.func,
    /**
     * Función que se llama cuando se libera un toque.
     */
    onPressOut: PropTypes.func,
    /**
     * Distancia adicional fuera de esta vista en la que un toque se considera una pulsación antes de que se active onPressOut.
     */
    pressRetentionOffset: PropTypes.shape({
        bottom: PropTypes.number,
        left: PropTypes.number,
        right: PropTypes.number,
        top: PropTypes.number,
    }),
    /**
     * Identificador para pruebas.
     */
    testID: PropTypes.string,
    /**
     * Identificador para accesibilidad.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad.
     */
    accessibilityHint: PropTypes.string,
};

const defaultProps = {
    style: DEFAULT_TYPES.style,
    delayLongPress: 500,
    disabled: false,
    hitSlop: {},
    onLongPress: DEFAULT_TYPES.func,
    onPress: DEFAULT_TYPES.func,
    onPressIn: DEFAULT_TYPES.func,
    onPressOut: DEFAULT_TYPES.func,
    pressRetentionOffset: {},
    testID: '',
    accessibilityLabel: '',
    accessibilityHint: '',
};

export { propTypes, defaultProps };
