// react-native-minerva/src/components/cardItemCollapsible/components/CardItemCollapsibleHeader.js
import React from 'react';
import PropTypes from 'prop-types';
import {
    CardItem,
    HelperStyle,
    Icon,
    Text,
    Validate,
    View,
} from '../../../index';
import Style from '../style';

const CardItemCollapsibleHeader = ({
    title,
    titleAccessibilityLabel,
    titleAccessibilityHint,
    titleTestID,
    iconName,
    iconLabel,
    iconAccessibilityHint,
    iconTestID,
}) => (
    <CardItem
        header
        style={[HelperStyle.bg_primary_background, Style.cardItemHeader]}
        accessible={false}
    >
        <View style={[HelperStyle.flx_i, HelperStyle.flx_row]}>
            {iconName ? (
                <Icon
                    style={[
                        HelperStyle.primary,
                        HelperStyle.fs_3,
                        HelperStyle.as_c,
                        HelperStyle.mr_2,
                    ]}
                    name={iconName}
                    accessible
                    accessibilityLabel={Validate.safeLabel(iconLabel)}
                    accessibilityHint={Validate.safeLabel(
                        iconAccessibilityHint,
                        iconLabel,
                    )}
                    testID={Validate.safeLabel(iconTestID, iconLabel)}
                />
            ) : null}
            <Text
                bold
                style={[
                    HelperStyle.gray_dark,
                    HelperStyle.mb_0,
                    HelperStyle.flx_grow,
                ]}
                accessibilityLabel={Validate.safeLabel(
                    titleAccessibilityLabel,
                    title,
                )}
                accessibilityHint={Validate.safeLabel(titleAccessibilityHint)}
                testID={Validate.safeLabel(titleTestID, title)}
            >
                {title}
            </Text>
        </View>
    </CardItem>
);

CardItemCollapsibleHeader.propTypes = {
    /**
     * Título del header.
     */
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    /**
     * Identificador de accesibilidad para title.
     */
    titleAccessibilityLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad que ayuda a los usuarios a comprender qué sucederá cuando realicen una acción.
     */
    titleAccessibilityHint: PropTypes.string,
    /**
     * Identificador de pruebas para title.
     */
    titleTestID: PropTypes.string,
    /**
     * Nombre del icono.
     */
    iconName: PropTypes.string,
    /**
     * Identificador de accesibilidad para el icon.
     */
    iconLabel: PropTypes.string,
    /**
     * Sugerencia de accesibilidad que ayuda a los usuarios a comprender qué sucederá cuando realicen una acción.
     */
    iconAccessibilityHint: PropTypes.string,
    /**
     * Identificador de pruebas para el icon.
     */
    iconTestID: PropTypes.string,
};

CardItemCollapsibleHeader.defaultProps = {
    titleAccessibilityLabel: '',
    titleAccessibilityHint: '',
    titleTestID: '',
    iconName: '',
    iconLabel: '',
    iconAccessibilityHint: '',
    iconTestID: '',
};

export default CardItemCollapsibleHeader;
