// react-native-minerva/src/components/listItemContactos/components/ListItemContactosRight.js
import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { DEFAULT_TYPES, TYPES } from '../../../constants/propTypes';
import { HelperStyle, Right, Text, Validate } from '../../../index';
import {
    listItemContactosPropTypes,
    listItemContactosDefaultProps,
} from '../type';

const ListItemContactosRight = ({
    accessibilityHint,
    accessibilityLabel,
    disabled,
    monto,
    styleMonto,
    styleRight,
    testID,
}) => (
    <Right style={[HelperStyle.ml_4, styleRight]}>
        <Text
            small
            number
            accessibilityLabel={Validate.safeLabel(accessibilityLabel, monto)}
            accessibilityHint={accessibilityHint}
            testID={Validate.safeLabel(testID, monto)}
            style={[
                styleMonto,
                Validate.style(disabled, HelperStyle.gray_lighter),
            ]}
        >
            {monto}
        </Text>
    </Right>
);

ListItemContactosRight.propTypes = {
    accessibilityHint: PropTypes.string,
    accessibilityLabel: PropTypes.string,
    disabled: listItemContactosPropTypes.disabled,
    monto: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    styleMonto: TYPES.style,
    styleRight: TYPES.style,
    testID: PropTypes.string,
};
ListItemContactosRight.defaultProps = {
    accessibilityHint: '',
    accessibilityLabel: '',
    disabled: listItemContactosDefaultProps.disabled,
    styleMonto: DEFAULT_TYPES.style,
    styleRight: DEFAULT_TYPES.style,
    testID: '',
};

export default memo(ListItemContactosRight);
