// react-native-minerva/src/components/cardTrampolin/style.js
import { StyleSheet } from 'react-native';
import { Scale, Spacing } from '../../index';

const style = StyleSheet.create({
    cardItem: {
        paddingTop: 0,
        paddingBottom: 0,
        paddingLeft: 0,
        paddingRight: 0,
        overflow: 'hidden',
        borderRadius: Scale(16),
    },
    cardLeft: {
        flex: 0.5,
        width: '100%',
        height: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        overflow: 'hidden',
    },
    cardRight: {
        flex: 0,
        height: '100%',
    },
    cardRightClose: {
        width: Scale(28),
        minWidth: Scale(28),
    },
    cardBody: {
        flex: 0.7,
        marginBottom: Spacing.spacer_4,
        marginTop: Spacing.spacer_4,
        paddingRight: Spacing.spacer_4,
        zIndex: 2,
    },
    wavesLeftMaskImageBackground: {
        position: 'absolute',
        width: '100%',
        height: '100%',
        overflow: 'hidden',
        zIndex: 1,
    },
    wavesLeftMaskImageStyle: {
        resizeMode: 'stretch',
        height: '100%',
        width: '100%',
    },
    iconClose: {
        position: 'absolute',
        top: Scale(8),
        right: Scale(8),
        zIndex: 2,
    },
    image: {
        width: '100%',
        height: '100%',
        resizeMode: 'cover',
    },
    imageBackground: {
        width: '100%',
        height: '100%',
        position: 'absolute',
        top: 0,
        zIndex: 0,
        overflow: 'hidden',
    },
});

export default style;
